{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "section": "header"
   },
   "source": [
    "<div style=\"text-align: center;\">\n",
    "<h1>The University of North Carolina at Chapel Hill</h1>\n",
    "<h1>Comp 555 BioAlgorithms - Spring 2018</h1>\n",
    "<h1 style=\"font-size: 250%;\">Problem Set #3 </h1>\n",
    "<h1>Issued Thursday, 3/1/2018; Due Thursday, 3/22/2018</h1>\n",
    "</div>\n",
    "\n",
    "**Homework Information:** Some of the problems are probably too long to be started the night before the due date, so plan accordingly. Late problem sets will be penalized by a factor of\t70.71% for each class meeting after the due date. Feel free to get help from others, but the work you submit in should be your own."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "section": "signature"
   },
   "outputs": [],
   "source": [
    "# Replace the following string values with the requested information\n",
    "class Student:\n",
    "    first = \"Mohamed\"\n",
    "    last = \"Lansari\"\n",
    "    onyen = \"mlans13\"\n",
    "    pid = \"730013359\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 1,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #1:** A file of *15-mers* simulating short reads from a genome can be downloaded [here](http://csbio.unc.edu/mcmillan/Comp555S18/data/kmers.txt). How many distinct nodes appear in the De Bruijn graph that represents these *15-mers* as edges? How many nodes are semi-balenced? How many nodes are balanced? How many are balanced with both in-degrees and out-degrees equal to 1?  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have 93673 in degrees and 93673 out-degrees\n",
      "We have 93674 vertices and 93925 edges\n",
      "The number of balanced nodes is 93672\n",
      "The number of semi-balanced nodes is 2\n",
      "The number of balanced nodes with in- and out-degree equal to one is 93522\n",
      "The length of the graph's eulerian path is 93926\n"
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "# Build a basic graph description\n",
    "class Graph:\n",
    "    def __init__(self, vlist=[]):\n",
    "        \"\"\" Initialize a Graph with an optional vertex list \"\"\"\n",
    "        self.index = {v:i for i,v in enumerate(vlist)}\n",
    "        self.vertex = {i:v for i,v in enumerate(vlist)}\n",
    "        self.edge = []\n",
    "        self.edgelabel = []\n",
    "    def addVertex(self, label):\n",
    "        \"\"\" Add a labeled vertex to the graph \"\"\"\n",
    "        index = len(self.index)\n",
    "        self.index[label] = index\n",
    "        self.vertex[index] = label\n",
    "    def addEdge(self, vsrc, vdst, label='', repeats=True):\n",
    "        \"\"\" Add a directed edge to the graph, with an optional label. \n",
    "        Repeated edges are distinct, unless repeats is set to False. \"\"\"\n",
    "        e = (self.index[vsrc], self.index[vdst])\n",
    "        if (repeats) or (e not in self.edge):\n",
    "            self.edge.append(e)\n",
    "            self.edgelabel.append(label)\n",
    "    def hamiltonianPath(self):\n",
    "        \"\"\" A Brute-force method for finding a Hamiltonian Path. \n",
    "        Basically, all possible N! paths are enumerated and checked\n",
    "        for edges. Since edges can be reused there are no distictions\n",
    "        made for *which* version of a repeated edge. \"\"\"\n",
    "        for path in itertools.permutations(sorted(self.index.values())):\n",
    "            for i in xrange(len(path)-1):\n",
    "                if ((path[i],path[i+1]) not in self.edge):\n",
    "                    break\n",
    "            else:\n",
    "                return [self.vertex[i] for i in path]\n",
    "        return []\n",
    "    def SearchTree(self, path, verticesLeft):\n",
    "        \"\"\" A recursive Branch-and-Bound Hamiltonian Path search. \n",
    "        Paths are extended one node at a time using only available\n",
    "        edges from the graph. \"\"\"\n",
    "        if (len(verticesLeft) == 0):\n",
    "            self.PathV2result = [self.vertex[i] for i in path]\n",
    "            return True\n",
    "        for v in verticesLeft:\n",
    "            if (len(path) == 0) or ((path[-1],v) in self.edge):\n",
    "                if self.SearchTree(path+[v], [r for r in verticesLeft if r != v]):\n",
    "                    return True\n",
    "        return False\n",
    "    def hamiltonianPathV2(self):\n",
    "        \"\"\" A wrapper function for invoking the Branch-and-Bound \n",
    "        Hamiltonian Path search. \"\"\"\n",
    "        self.PathV2result = []\n",
    "        self.SearchTree([],sorted(self.index.values()))                \n",
    "        return self.PathV2result\n",
    "    def degrees(self):\n",
    "        \"\"\" Returns two dictionaries with the inDegree and outDegree\n",
    "        of each node from the graph. \"\"\"\n",
    "        inDegree = {}\n",
    "        outDegree = {}\n",
    "        for src, dst in self.edge:\n",
    "            outDegree[src] = outDegree.get(src, 0) + 1\n",
    "            inDegree[dst] = inDegree.get(dst, 0) + 1\n",
    "        return inDegree, outDegree\n",
    "    def verifyAndGetStart(self):\n",
    "        inDegree, outDegree = self.degrees()\n",
    "        start = 0\n",
    "        end = 0\n",
    "        for vert in self.vertex.iterkeys():\n",
    "            ins = inDegree.get(vert,0)\n",
    "            outs = outDegree.get(vert,0)\n",
    "            if (ins == outs):\n",
    "                continue\n",
    "            elif (ins - outs == 1):\n",
    "                end = vert\n",
    "            elif (outs - ins == 1):\n",
    "                start = vert\n",
    "            else:\n",
    "                start, end = -1, -1\n",
    "                break\n",
    "        if (start >= 0) and (end >= 0):\n",
    "            return start\n",
    "        else:\n",
    "            return -1\n",
    "    def eulerianPath(self):\n",
    "        graph = [(src,dst) for src,dst in self.edge]\n",
    "        currentVertex = self.verifyAndGetStart()\n",
    "        path = [currentVertex]\n",
    "        # \"next\" is where vertices get inserted into our tour\n",
    "        # it starts at the end (i.e. it is the same as appending),\n",
    "        # but later \"side-trips\" will insert in the middle\n",
    "        next = 1\n",
    "        while len(graph) > 0:\n",
    "            for edge in graph:\n",
    "                if (edge[0] == currentVertex):\n",
    "                    currentVertex = edge[1]\n",
    "                    graph.remove(edge)\n",
    "                    path.insert(next, currentVertex)\n",
    "                    next += 1\n",
    "                    break\n",
    "            else:\n",
    "                for edge in graph:\n",
    "                    try:\n",
    "                        next = path.index(edge[0]) + 1\n",
    "                        currentVertex = edge[0]\n",
    "                        break\n",
    "                    except ValueError:\n",
    "                        continue\n",
    "                else:\n",
    "                    print \"There is no path!\"\n",
    "                    return False\n",
    "        return path\n",
    "    def eulerEdges(self, path):\n",
    "        edgeId = {}\n",
    "        for i in xrange(len(self.edge)):\n",
    "            edgeId[self.edge[i]] = edgeId.get(self.edge[i], []) + [i]\n",
    "        edgeList = []\n",
    "        for i in xrange(len(path)-1):\n",
    "            edgeList.append(self.edgelabel[edgeId[path[i],path[i+1]].pop()])            \n",
    "        return edgeList\n",
    "    def render(self, highlightPath=[]):\n",
    "        \"\"\" Outputs a version of the graph that can be rendered\n",
    "        using graphviz tools (http://www.graphviz.org/).\"\"\"\n",
    "        edgeId = {}\n",
    "        for i in xrange(len(self.edge)):\n",
    "            edgeId[self.edge[i]] = edgeId.get(self.edge[i], []) + [i]\n",
    "        edgeSet = set()\n",
    "        for i in xrange(len(highlightPath)-1):\n",
    "            src = self.index[highlightPath[i]]\n",
    "            dst = self.index[highlightPath[i+1]]\n",
    "            edgeSet.add(edgeId[src,dst].pop())\n",
    "        result = ''\n",
    "        result += 'digraph {\\n'\n",
    "        result += '   graph [nodesep=2, size=\"10,10\"];\\n'\n",
    "        for index, label in self.vertex.iteritems():\n",
    "            result += '    N%d [shape=\"box\", style=\"rounded\", label=\"%s\"];\\n' % (index, label)\n",
    "        for i, e in enumerate(self.edge):\n",
    "            src, dst = e\n",
    "            result += '    N%d -> N%d' % (src, dst)\n",
    "            label = self.edgelabel[i]\n",
    "            if (len(label) > 0):\n",
    "                if (i in edgeSet):\n",
    "                    result += ' [label=\"%s\", penwidth=3.0]' % (label)\n",
    "                else:\n",
    "                    result += ' [label=\"%s\"]' % (label)\n",
    "            elif (i in edgeSet):\n",
    "                result += ' [penwidth=3.0]'                \n",
    "            result += ';\\n'                \n",
    "        result += '    overlap=false;\\n'\n",
    "        result += '}\\n'\n",
    "        return result\n",
    "    \n",
    "# function to read a file of k-mer edges, and use them to describe a graph\n",
    "def graphFromEdgeFile(filename, k):\n",
    "    grph = Graph()\n",
    "    vert = set()\n",
    "    edges = []\n",
    "    # read in the file\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            vert.add(line[0:k - 1])\n",
    "            vert.add(line[1:k])\n",
    "            edges.append(line[0:k])\n",
    "    \n",
    "    for v in vert:\n",
    "        grph.addVertex(v)\n",
    "    for e in edges:\n",
    "        grph.addEdge(e[0:k - 1], e[1:k], e)\n",
    "            \n",
    "    return grph, edges, vert\n",
    "\n",
    "# lets assess the degrees\n",
    "g, edges, verts = graphFromEdgeFile(\"kmers.txt\", 15)\n",
    "inDeg, outDeg = g.degrees()\n",
    "\n",
    "# information about our graph\n",
    "print \"We have\", len(inDeg), \"in degrees and\", len(outDeg), \"out-degrees\"\n",
    "print \"We have\", len(verts), \"vertices and\", len(edges), \"edges\"\n",
    "\n",
    "# perform in- versus out-degree comparison\n",
    "nBal = 0\n",
    "nSemi = 0\n",
    "inOutUno= 0\n",
    "\n",
    "# generate the set of both in-degree and out-degree keys\n",
    "keysTot = set()\n",
    "for k in inDeg.keys():\n",
    "    keysTot.add(k)\n",
    "for k in outDeg.keys():\n",
    "    keysTot.add(k)\n",
    "    \n",
    "# calculate the various styles of node\n",
    "for e in keysTot:\n",
    "    if inDeg.get(e, 0) == outDeg.get(e, 0):\n",
    "        nBal += 1\n",
    "        if inDeg.get(e, 0) == 1:\n",
    "            inOutUno += 1\n",
    "    elif abs(inDeg.get(e, 0) - outDeg.get(e, 0)) == 1:\n",
    "        nSemi += 1\n",
    "\n",
    "# print our results\n",
    "print \"The number of balanced nodes is\", nBal\n",
    "print \"The number of semi-balanced nodes is\", nSemi\n",
    "print \"The number of balanced nodes with in- and out-degree equal to one is\", inOutUno\n",
    "\n",
    "# calculate the eulerian path only once\n",
    "ePath = g.eulerianPath()\n",
    "print \"The length of the graph's eulerian path is\", len(ePath)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 1,
    "section": "answer"
   },
   "source": [
    "93674 total nodes\n",
    "93672 balanced nodes\n",
    "2 semi-balanced nodes\n",
    "93522 balanced nodes with both in- and out-degrees of 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 2,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #2:** What is the length of the Eulerian path that can be constructed in the De Bruijn graph described in Problem #1? How does the resulting constructed sequence compare to the plasmid sequence of [*Salmonella Typhimurium*](http://csbio.unc.edu/mcmillan/Comp555S18/data/SalmonellaTyphimurium.fa). (Hint: one method of comparison is consider how many k-mers the two sequences share. Of course you can compare where k=15, but consider what is the smallest value of k for which the two sequences differ in their set of k-mers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93926\n",
      "The length of eTrans is 93939\n",
      "A slice of eTrans is GTATCACAACGGTATTTTTGCCACCAGTCATTATGAAGGCGCGCTCACCAGTGAAGATATCAGTAGTGCGGCGGCCGCCGCCCTGAAAGGGCATAAAATC\n",
      "salDNA from 0 to 100 is GAGTGAACGGATGAAACAGAAAGACCGTCTGTACGGCGTGGCACCGGCCTTACCCCGATTGCAGGCTGTGAAGCTAGGCCGCAGGTCCGCTATGTGTTTA\n",
      "the current k val is 3\n",
      "The length of gMers is 64 and the length of sMers is 64\n",
      "The delta between them is set([])\n",
      "the current k val is 4\n",
      "The length of gMers is 256 and the length of sMers is 256\n",
      "The delta between them is set([])\n",
      "the current k val is 5\n",
      "The length of gMers is 1024 and the length of sMers is 1024\n",
      "The delta between them is set([])\n",
      "the current k val is 6\n",
      "The length of gMers is 4071 and the length of sMers is 4071\n",
      "The delta between them is set([])\n",
      "the current k val is 7\n",
      "The length of gMers is 15230 and the length of sMers is 15230\n",
      "The delta between them is set([])\n",
      "the current k val is 8\n",
      "The length of gMers is 41540 and the length of sMers is 41540\n",
      "The delta between them is set([])\n",
      "the current k val is 9\n",
      "The length of gMers is 69982 and the length of sMers is 69982\n",
      "The delta between them is set([])\n",
      "the current k val is 10\n",
      "The length of gMers is 85411 and the length of sMers is 85411\n",
      "The delta between them is set([])\n",
      "the current k val is 11\n",
      "The length of gMers is 91119 and the length of sMers is 91119\n",
      "The delta between them is set([])\n",
      "the current k val is 12\n",
      "The length of gMers is 92937 and the length of sMers is 92937\n",
      "The delta between them is set([])\n",
      "the current k val is 13\n",
      "The length of gMers is 93497 and the length of sMers is 93497\n",
      "The delta between them is set([])\n",
      "the current k val is 14\n",
      "The length of gMers is 93674 and the length of sMers is 93674\n",
      "The delta between them is set([])\n",
      "the current k val is 15\n",
      "The length of gMers is 93741 and the length of sMers is 93741\n",
      "The delta between them is set([])\n",
      "the current k val is 16\n",
      "The length of gMers is 93765 and the length of sMers is 93765\n",
      "The delta between them is set(['GTGTTTTTCATCATAA', 'CGGGATGTGGTGAACG', 'GTGTGAAATATTTCCC', 'TATGCCATTCTGAACA', 'AATGCCATTCTGAACC', 'TGACCCGGATGGCGGG', 'GCCATCACCCGCGCCT', 'GGCTCCAGGTCATCCA', 'AGGGATGTGGTGAACC', 'AAGCCAGCCCGTCAGG', 'GCTCAATTCTCATGAC', 'CATCCGTGACGGGGAG', 'TGCTGCCGGAGAGTGA', 'CAATGCCATTCTGAAC', 'TCAGGACAAATACACC', 'CCAGAAAAAAAAGCCG', 'GCAGGACAAATACACG', 'CGCTGCCGGAGAGTGT', 'TCCATCACCCGCGCCA', 'ACTCAATTCTCATGAG', 'TATAAAATAAATTTCC', 'AAAAATCAGCAGCGGC', 'CAGCCAGCCCGTCAGC', 'CAAAATCAGCAGCGGA', 'CGCCGACCTGCTGGCA', 'GGGCGCTGGATACGGA', 'GGACCCGGATGGCGGT', 'ATGTTTTTCATCATAT', 'AATAAAATAAATTTCA', 'TATCCGTGACGGGGAA', 'TGGCGCTGGATACGGC', 'GAATGCCATTCTGAAA', 'TTGTGAAATATTTCCG', 'GGATGCGCTGCTGGTT', 'AGATGCGCTGCTGGTC', 'GCAGAAAAAAAAGCCC', 'ACGGGGTGTACCTGGG', 'CGGGCTGATACAGCCG', 'AGTTTCGTGAATGTAA', 'GGGGCTGATACAGCCA', 'TGTTTCGTGAATGTAC', 'GCGGGGTGTACCTGGA', 'TGCCGACCTGCTGGCG', 'AGCTCCAGGTCATCCG'])\n"
     ]
    }
   ],
   "source": [
    "# opening fasta\n",
    "# we need to begin by loading the genome\n",
    "def loadFasta(filename):\n",
    "    \"\"\" Parses a classically formatted and possibly \n",
    "        compressed FASTA file into a list of headers \n",
    "        and fragment sequences for each sequence contained\"\"\"\n",
    "    if (filename.endswith(\".gz\")):\n",
    "        fp = gzip.open(filename, 'rb')\n",
    "    else:\n",
    "        fp = open(filename, 'r')\n",
    "    # split at headers\n",
    "    data = fp.read().split('>')\n",
    "    fp.close()\n",
    "    # ignore whatever appears before the 1st header\n",
    "    data.pop(0)     \n",
    "    headers = []\n",
    "    sequences = []\n",
    "    for sequence in data:\n",
    "        lines = sequence.split('\\n')\n",
    "        headers.append(lines.pop(0))\n",
    "        # add an extra \"+\" to make string \"1-referenced\"\n",
    "        sequences.append('+' + ''.join(lines))\n",
    "    return (headers, sequences)\n",
    "\n",
    "# start with the generated eulerian path\n",
    "print len(ePath)\n",
    "\n",
    "# build the actual path\n",
    "eTrans = g.vertex[ePath[0]]\n",
    "for n in xrange(1, len(ePath)):\n",
    "    eTrans += g.vertex[ePath[n]][13]\n",
    "\n",
    "print \"The length of eTrans is\", len(eTrans)\n",
    "print \"A slice of eTrans is\", eTrans[200:300]\n",
    "\n",
    "# now we need to calculate the smallest value of k for which the two sequences differ in their set of k-mers\n",
    "headers, sequences = loadFasta(\"SalmonellaTyphimurium.fa\")\n",
    "salDNA = sequences[1][1:]   # this is the genome in question\n",
    "\n",
    "print \"salDNA from 0 to 100 is\", salDNA[0:100]\n",
    "\n",
    "# compile all the k-mers of each sequence, both the path and from the salmonella\n",
    "# start at 15?\n",
    "gMers = {}   # kmers for graph\n",
    "sMers = {}   # kmers for salmonella\n",
    "for k in xrange(3, len(eTrans)):\n",
    "    # current k val\n",
    "    print \"the current k val is\", k\n",
    "    \n",
    "    # generate the kmer sets\n",
    "    gMers[k] = set()\n",
    "    sMers[k] = set()\n",
    "    \n",
    "    # generate the kmer lists for each\n",
    "    for i in xrange(len(eTrans) - k + 1):\n",
    "        gMers[k].add(str(eTrans[i:i + k]))\n",
    "    for i in xrange(len(salDNA) - k + 1):\n",
    "        sMers[k].add(str(salDNA[i:i + k]))\n",
    "        \n",
    "    # check for the delta between the sets\n",
    "    print \"The length of gMers is\", len(gMers[k]), \"and the length of sMers is\", len(sMers[k])\n",
    "    print \"The delta between them is\", sMers[k] - gMers[k]\n",
    "                \n",
    "    if len(sMers[k] - gMers[k]) != 0:\n",
    "        break\n",
    "    "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 2,
    "section": "answer"
   },
   "source": [
    "The length of the Eulerian path is 93926 vertices, or 93925 edges.  This correlates to an actual length of the transcribed genome of 93939.\n",
    "Starting at k value of 16, the plasmid sequence of the salmonella genome differs from the genome we've mapped to our graph.  We move through from the lowest amounts increasing, and check for deltas between the sets of k-mers within each string, until we get to 16-mers, where the salmonella plasmid sequence has 16-mers that the graph genome does not contain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 3,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #3:** You will find a BWT of the primary \"genome\" sequence of *Salmonella Typhimurium* [here](http://csbio.unc.edu/mcmillan/Comp555S18/data/SalmonellaTyphimurium.bwt). This BWT is compressed as follows: Any run of two or more repeated characters is prefixed with an ASCII-encoded number followed by the character. For example the string \"AAACGGTTTTTTTTTT\" would be encoded as the string \"3AC2G10T\". What is the compression ratio of this BWT? Where the compression ration is given by: \n",
    "\n",
    "$$\\frac{len(compressed BWT)}{len(sequence)}$$\n",
    "\n",
    "What is the average run-length in the BWT (consider characters without runs as run-lengths of 1)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The original length of the bwt is 4397999\n",
      "The length of the expanded bwt is 4857433\n",
      "This means that the compression ratio is 0.905416297044\n",
      "The average read length of the bwt is 1.40828468415\n"
     ]
    }
   ],
   "source": [
    "# USEFUL\n",
    "codons = [\"A\", \"T\", \"C\", \"G\", \"$\"]\n",
    "\n",
    "# we begin by loading in the BWT\n",
    "bwt = \"\"\n",
    "with open(\"SalmonellaTyphimurium.bwt\", 'r') as f:\n",
    "    bwt = f.read().replace(\"\\n\", '')\n",
    "\n",
    "# function to parse the bwt's full length\n",
    "def getExpansion(seq):\n",
    "    ind = 0\n",
    "    expLen = 0\n",
    "    readLens = []\n",
    "    while ind < len(seq):\n",
    "        num = \"\"\n",
    "        while seq[ind] not in codons:\n",
    "            num = num + seq[ind]\n",
    "            ind += 1\n",
    "        if num != \"\":\n",
    "            expLen += int(num)\n",
    "            readLens.append(int(num))\n",
    "        else:\n",
    "            expLen += 1\n",
    "            readLens.append(1)\n",
    "        ind += 1\n",
    "    \n",
    "    return expLen, readLens\n",
    "\n",
    "# get the expansion of the bwt\n",
    "expansion, reads = getExpansion(bwt)\n",
    "\n",
    "# calculate the average read length\n",
    "avg = 0.0\n",
    "for n in reads:\n",
    "    avg += n\n",
    "avg /= len(reads)\n",
    "\n",
    "print \"The original length of the bwt is\", len(bwt)\n",
    "print \"The length of the expanded bwt is\", expansion\n",
    "print \"This means that the compression ratio is\", (len(bwt) / float(expansion))\n",
    "print \"The average read length of the bwt is\", avg"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 3,
    "section": "answer"
   },
   "source": [
    "Our compression ratio is 0.905416277572\n",
    "The average read length of the bwt is 1.40828480252"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 4,
    "section": "problem"
   },
   "source": [
    "**Problem #4:** Uncompress the BWT from in Problem #3, and use code given in class to find how many times the substring \"ATGACAACGC\" and its reverse complement \"GCGTTGTCAT\" appear in the *Salmonella Typhimurium* genome. Repeat this for the substring \"ATGACAACGCAT\" and its reverse complement \"ATGCGTTGTCAT\". Note these sequences are the first few bases of the *HolC* gene that you searched for in Problem Set #2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have 4 occurences of ATGACAACGC\n",
      "We have 13 occurences of GCGTTGTCAT, its reverse complement\n",
      "We have 1 occurences of ATGACAACGCAT\n",
      "We have 1 occurences of ATGCGTTGTCAT, its reverse complement\n"
     ]
    }
   ],
   "source": [
    "# create a bwt transcription function\n",
    "def getBWTTranscription(seq):\n",
    "    ind = 0\n",
    "    transcription = \"\"\n",
    "    \n",
    "    # iterate through the entire bwt\n",
    "    while ind < len(seq):\n",
    "        num = \"\"\n",
    "        while seq[ind] not in codons:\n",
    "            num += seq[ind]\n",
    "            ind += 1\n",
    "        if num != \"\":\n",
    "            transcription += (seq[ind] * int(num))\n",
    "        else:\n",
    "            transcription += seq[ind]\n",
    "        ind += 1\n",
    "    return transcription\n",
    "\n",
    "# method to generate an FMIndex\n",
    "def FMIndex(bwt):\n",
    "    fm = [{c: 0 for c in bwt}]\n",
    "    for c in bwt:\n",
    "        row = {symbol: count + 1 if (symbol == c) else count for symbol, count in fm[-1].iteritems()}\n",
    "        fm.append(row)\n",
    "    offset = {}\n",
    "    N = 0\n",
    "    for symbol in sorted(row.keys()):\n",
    "        offset[symbol] = N\n",
    "        N += row[symbol]\n",
    "    return fm, offset\n",
    "\n",
    "# method used to search bwt for substrings\n",
    "def findBWT(pattern, FMIndex, Offset):\n",
    "    lo = 0\n",
    "    hi = len(FMIndex) - 1\n",
    "    for symbol in reversed(pattern):\n",
    "        lo = Offset[symbol] + FMIndex[lo][symbol]\n",
    "        hi = Offset[symbol] + FMIndex[hi][symbol]\n",
    "    return lo, hi\n",
    "\n",
    "# get the transcription\n",
    "trans = getBWTTranscription(bwt)\n",
    "\n",
    "# generate an FMIndex for the BWT\n",
    "fm, offset = FMIndex(trans)\n",
    "\n",
    "# search for all occurences of \"ATGACAACGC\" and its reverse complement\n",
    "lo1, hi1 = findBWT(\"ATGACAACGC\", fm, offset)\n",
    "print \"We have\", hi1 - lo1, \"occurences of ATGACAACGC\"\n",
    "lo2, hi2 = findBWT(\"GCGTTGTCAT\", fm, offset)\n",
    "print \"We have\", hi2 - lo2, \"occurences of GCGTTGTCAT, its reverse complement\"\n",
    "\n",
    "# search for all occurences of \"ATGACAACGCAT\" and its reverse complement\n",
    "lo3, hi3 = findBWT(\"ATGACAACGCAT\", fm, offset)\n",
    "print \"We have\", hi3 - lo3, \"occurences of ATGACAACGCAT\"\n",
    "lo4, hi4 = findBWT(\"ATGACAACGCAT\"[::-1], fm, offset)\n",
    "print \"We have\", hi4 - lo4, \"occurences of ATGCGTTGTCAT, its reverse complement\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "collapsed": true,
    "number": 4,
    "section": "answer"
   },
   "source": [
    "We have 4 occurences of ATGACAACGC\n",
    "We have 13 occurences of GCGTTGTCAT, its reverse complement\n",
    "We have 1 occurence of ATGACAACGCAT\n",
    "We have 1 occurence of ATGCGTTGTCAT, its reverse complement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 5,
    "section": "problem"
   },
   "source": [
    "**Problem #5: Programming Problem** \n",
    "\n",
    "In class we discussed an algortihm to produce a BWT from a suffix array. In this problem you are asked to write code to do the inverse-- ***produce a suffix array from a BWT***. In the space provided below write your function and test it by finding the genomic indices for all of the substrings than you reported for Problem #4. (Hint: The first implict suffix in the BWT begins with '$', which is the last character of the original string)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "number": 5,
    "section": "answer"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The suffix array is of length 4857433\n",
      "ATGACAACGC is at 1006271 to 1006275 [2866253, 2253887, 299096, 4756596]\n",
      "GCGTTGTCAT is at [3840958, 4244113, 1945357, 3973758, 2706228, 1160902, 3509441, 3113230, 4720340, 1357533, 2939381, 3879426, 3884275]\n",
      "ATGACAACGCAT is at [299096]\n",
      "ATGCGTTGTCAT is at [3899847]\n"
     ]
    }
   ],
   "source": [
    "# Enter your code here\n",
    "codons = [\"A\", \"T\", \"C\", \"G\", \"$\"]\n",
    "\n",
    "# we begin by loading in the BWT\n",
    "def loadBWT():\n",
    "    bwt = \"\"\n",
    "    with open(\"SalmonellaTyphimurium.bwt\", 'r') as f:\n",
    "        bwt = f.read().replace(\"\\n\", '')\n",
    "    return bwt\n",
    "\n",
    "\n",
    "def FMIndex(bwt):\n",
    "    fm = [{c: 0 for c in bwt}]\n",
    "    for c in bwt:\n",
    "        row = {symbol: count + 1 if (symbol == c) else count for symbol, count in fm[-1].iteritems()}\n",
    "        fm.append(row)\n",
    "    offset = {}\n",
    "    N = 0\n",
    "    for symbol in sorted(row.keys()):\n",
    "        offset[symbol] = N\n",
    "        N += row[symbol]\n",
    "    return fm, offset\n",
    "\n",
    "def getBWTTranscription(seq):\n",
    "    ind = 0\n",
    "    transcription = \"\"\n",
    "    \n",
    "    # iterate through the entire bwt\n",
    "    while ind < len(seq):\n",
    "        num = \"\"\n",
    "        while seq[ind] not in codons:\n",
    "            num += seq[ind]\n",
    "            ind += 1\n",
    "        if num != \"\":\n",
    "            transcription += (seq[ind] * int(num))\n",
    "        else:\n",
    "            transcription += seq[ind]\n",
    "        ind += 1\n",
    "    return transcription\n",
    "\n",
    "\n",
    "# THIS IS THE METHOD THAT IMPLEMENTS RETRIEVING A SUFFIX ARRAY FROM A BWT\n",
    "def retrieveSuffixArray(bwt):\n",
    "    # begin by generating the fm index and the offsets of the bwt inputted\n",
    "    fm, offset = FMIndex(bwt)\n",
    "    \n",
    "    # generate a list to fill of the correct length\n",
    "    arr = [0 for c in bwt]\n",
    "    \n",
    "    # we start by find the first item in the suffix array, the terminator\n",
    "    predec = offset[\"$\"] + fm[bwt.index(\"$\")][\"$\"]\n",
    "    for i in xrange(1, len(bwt)):\n",
    "        arr[predec] = len(bwt) - i\n",
    "        predec = offset[bwt[predec]] + fm[predec][bwt[predec]]\n",
    "    \n",
    "    return arr\n",
    "\n",
    "# fetch the suffix array and the original text\n",
    "# arr, text = retrieveSuffixArray(getBWTTranscription(loadBWT()))\n",
    "arr = retrieveSuffixArray(getBWTTranscription(loadBWT()))\n",
    "\n",
    "# finding the genomic indices of ATGACAACGC, GCGTTGTCAT, ATGACAACGCAT, ATGCGTTGTCAT\n",
    "# this is the testing\n",
    "print \"ATGACAACGC is at\", arr[lo1:hi1]\n",
    "print \"GCGTTGTCAT is at\", arr[lo2:hi2]\n",
    "print \"ATGACAACGCAT is at\", arr[lo3:hi3]\n",
    "print \"ATGCGTTGTCAT is at\", arr[lo4:hi4]\n",
    "\n",
    "# the results are\n",
    "#\n",
    "# ATGACAACGC is at 1006271 to 1006275 [2866253, 2253887, 299096, 4756596]\n",
    "# GCGTTGTCAT is at [3840958, 4244113, 1945357, 3973758, 2706228, 1160902, 3509441, 3113230, 4720340, \n",
    "#                   1357533, 2939381, 3879426, 3884275]\n",
    "# ATGACAACGCAT is at [299096]\n",
    "# ATGCGTTGTCAT is at [3899847]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "section": "submit"
   },
   "source": [
    "### Click [here to submit](http://csbio.unc.edu/mcmillan/index.py?run=PS.upload) your completed problem set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
