{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "section": "header"
   },
   "source": [
    "<div style=\"text-align: center;\">\n",
    "<h1>The University of North Carolina at Chapel Hill</h1>\n",
    "<h1>Comp 555 BioAlgorithms - Spring 2018</h1>\n",
    "<h1 style=\"font-size: 250%;\">Problem Set #2 </h1>\n",
    "<h1>Issued Thursday, 2/15/2018; Due Thursday, 3/1/2018</h1>\n",
    "</div>\n",
    "\n",
    "**Homework Information:** Some of the problems are probably too long to complete the night before the due date, so plan accordingly. Late problem sets will be penalized by a factor of\t70.71% for each class meeting after the due date. Feel free to get help from others, but the work you submit in should be your own."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "section": "signature"
   },
   "outputs": [],
   "source": [
    "# Replace the following string values with the requested information\n",
    "class Student:\n",
    "    first = \"\"\n",
    "    last = \"\"\n",
    "    onyen = \"\"\n",
    "    pid = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 1,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #1:** Consider the table of integer molecular weights for amino acid residues that we have used in class and is repeated below.\n",
    "\n",
    "        Daltons = { \n",
    "            'A':  71, 'C': 103, 'D': 115, 'E': 129,\n",
    "            'F': 147, 'G':  57, 'H': 137, 'I': 113,\n",
    "            'K': 128, 'L': 113, 'M': 131, 'N': 114,\n",
    "            'P':  97, 'Q': 128, 'R': 156, 'S':  87,\n",
    "            'T': 101, 'V':  99, 'W': 186, 'Y': 163 \n",
    "        }\n",
    "\n",
    "Which peptide *combinations* have a total molecular weight matching the peptide \"YMKN\"? A *combination* means that the order of the peptides is unimportant (i.e. \"AKS\" is the same combination as any of \"ASK\", \"KAS\", \"KSA\", \"SAK\", and \"SKA\"). When reporting your answer you should lexigraphically sort the resides within each combination (\"AKS\" is the sorted version of the previous example). Furthermore, the list of combinations should also be lexigraphically sorted (The python sorted() method will do this for you)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "536\n",
      "[['A'], ['C'], ['E'], ['D'], ['G'], ['F'], ['I'], ['H'], ['K'], ['M'], ['L'], ['N'], ['Q'], ['P'], ['S'], ['R'], ['T'], ['W'], ['V'], ['Y']]\n",
      "213\n",
      "[['A', 'A', 'A', 'A', 'D', 'H'], ['A', 'A', 'A', 'C', 'G', 'Y'], ['A', 'A', 'A', 'E', 'G', 'H'], ['A', 'A', 'A', 'E', 'P', 'P'], ['A', 'A', 'A', 'H', 'S', 'V'], ['A', 'A', 'A', 'H', 'W'], ['A', 'A', 'A', 'I', 'I', 'P'], ['A', 'A', 'A', 'I', 'L', 'P'], ['A', 'A', 'A', 'L', 'L', 'P'], ['A', 'A', 'C', 'C', 'G', 'M'], ['A', 'A', 'C', 'C', 'S', 'T'], ['A', 'A', 'C', 'F', 'G', 'S'], ['A', 'A', 'C', 'G', 'H', 'P'], ['A', 'A', 'C', 'K', 'Y'], ['A', 'A', 'C', 'P', 'P', 'P'], ['A', 'A', 'C', 'Q', 'Y'], ['A', 'A', 'E', 'H', 'K'], ['A', 'A', 'E', 'H', 'Q'], ['A', 'A', 'G', 'H', 'I', 'S'], ['A', 'A', 'G', 'H', 'L', 'S'], ['A', 'A', 'G', 'H', 'T', 'V'], ['A', 'A', 'G', 'S', 'S', 'Y'], ['A', 'A', 'H', 'R', 'T'], ['A', 'A', 'I', 'P', 'P', 'S'], ['A', 'A', 'L', 'P', 'P', 'S'], ['A', 'A', 'P', 'P', 'T', 'V'], ['A', 'A', 'P', 'V', 'V', 'V'], ['A', 'C', 'C', 'C', 'G', 'V'], ['A', 'C', 'C', 'C', 'R'], ['A', 'C', 'C', 'D', 'G', 'S'], ['A', 'C', 'C', 'G', 'T', 'T'], ['A', 'C', 'C', 'K', 'M'], ['A', 'C', 'C', 'M', 'Q'], ['A', 'C', 'F', 'G', 'G', 'T'], ['A', 'C', 'F', 'K', 'S'], ['A', 'C', 'F', 'N', 'T'], ['A', 'C', 'F', 'Q', 'S'], ['A', 'C', 'G', 'M', 'S', 'S'], ['A', 'C', 'H', 'K', 'P'], ['A', 'C', 'H', 'P', 'Q'], ['A', 'C', 'S', 'S', 'S', 'T'], ['A', 'D', 'G', 'G', 'H', 'V'], ['A', 'D', 'G', 'H', 'R'], ['A', 'D', 'G', 'P', 'P', 'V'], ['A', 'D', 'H', 'N', 'V'], ['A', 'D', 'P', 'P', 'R'], ['A', 'F', 'F', 'G', 'G', 'G'], ['A', 'F', 'F', 'G', 'N'], ['A', 'F', 'G', 'S', 'S', 'S'], ['A', 'G', 'G', 'G', 'M', 'Y'], ['A', 'G', 'G', 'H', 'I', 'T'], ['A', 'G', 'G', 'H', 'L', 'T'], ['A', 'G', 'G', 'S', 'T', 'Y'], ['A', 'G', 'H', 'P', 'S', 'S'], ['A', 'G', 'I', 'P', 'P', 'T'], ['A', 'G', 'I', 'P', 'V', 'V'], ['A', 'G', 'L', 'P', 'P', 'T'], ['A', 'G', 'L', 'P', 'V', 'V'], ['A', 'G', 'M', 'N', 'Y'], ['A', 'H', 'I', 'K', 'S'], ['A', 'H', 'I', 'N', 'T'], ['A', 'H', 'I', 'Q', 'S'], ['A', 'H', 'K', 'L', 'S'], ['A', 'H', 'K', 'T', 'V'], ['A', 'H', 'L', 'N', 'T'], ['A', 'H', 'L', 'Q', 'S'], ['A', 'H', 'Q', 'T', 'V'], ['A', 'I', 'P', 'R', 'V'], ['A', 'K', 'S', 'S', 'Y'], ['A', 'L', 'P', 'R', 'V'], ['A', 'N', 'S', 'T', 'Y'], ['A', 'P', 'P', 'P', 'S', 'S'], ['A', 'Q', 'S', 'S', 'Y'], ['C', 'C', 'C', 'G', 'G', 'I'], ['C', 'C', 'C', 'G', 'G', 'L'], ['C', 'C', 'C', 'I', 'N'], ['C', 'C', 'C', 'K', 'V'], ['C', 'C', 'C', 'L', 'N'], ['C', 'C', 'C', 'Q', 'V'], ['C', 'C', 'D', 'G', 'G', 'T'], ['C', 'C', 'D', 'K', 'S'], ['C', 'C', 'D', 'N', 'T'], ['C', 'C', 'D', 'Q', 'S'], ['C', 'C', 'E', 'G', 'G', 'S'], ['C', 'C', 'E', 'N', 'S'], ['C', 'C', 'G', 'S', 'S', 'V'], ['C', 'C', 'G', 'S', 'W'], ['C', 'C', 'K', 'T', 'T'], ['C', 'C', 'Q', 'T', 'T'], ['C', 'C', 'R', 'S', 'S'], ['C', 'D', 'F', 'G', 'G', 'G'], ['C', 'D', 'F', 'G', 'N'], ['C', 'D', 'G', 'S', 'S', 'S'], ['C', 'F', 'G', 'K', 'T'], ['C', 'F', 'G', 'Q', 'T'], ['C', 'G', 'G', 'G', 'M', 'M'], ['C', 'G', 'G', 'G', 'V', 'Y'], ['C', 'G', 'G', 'M', 'S', 'T'], ['C', 'G', 'G', 'R', 'Y'], ['C', 'G', 'M', 'M', 'N'], ['C', 'G', 'N', 'V', 'Y'], ['C', 'G', 'S', 'S', 'T', 'T'], ['C', 'K', 'M', 'S', 'S'], ['C', 'M', 'N', 'S', 'T'], ['C', 'M', 'Q', 'S', 'S'], ['C', 'N', 'R', 'Y'], ['D', 'E', 'E', 'Y'], ['D', 'F', 'H', 'H'], ['D', 'G', 'G', 'G', 'H', 'I'], ['D', 'G', 'G', 'G', 'H', 'L'], ['D', 'G', 'G', 'G', 'S', 'Y'], ['D', 'G', 'G', 'I', 'P', 'P'], ['D', 'G', 'G', 'L', 'P', 'P'], ['D', 'G', 'H', 'I', 'N'], ['D', 'G', 'H', 'K', 'V'], ['D', 'G', 'H', 'L', 'N'], ['D', 'G', 'H', 'Q', 'V'], ['D', 'G', 'N', 'S', 'Y'], ['D', 'H', 'K', 'R'], ['D', 'H', 'Q', 'R'], ['D', 'I', 'N', 'P', 'P'], ['D', 'K', 'P', 'P', 'V'], ['D', 'L', 'N', 'P', 'P'], ['D', 'P', 'P', 'Q', 'V'], ['E', 'E', 'F', 'M'], ['E', 'F', 'F', 'I'], ['E', 'F', 'F', 'L'], ['E', 'F', 'P', 'Y'], ['E', 'G', 'G', 'G', 'H', 'V'], ['E', 'G', 'G', 'H', 'R'], ['E', 'G', 'G', 'P', 'P', 'V'], ['E', 'G', 'H', 'N', 'V'], ['E', 'G', 'P', 'P', 'R'], ['E', 'H', 'N', 'R'], ['E', 'I', 'M', 'Y'], ['E', 'L', 'M', 'Y'], ['E', 'N', 'P', 'P', 'V'], ['F', 'F', 'G', 'G', 'K'], ['F', 'F', 'G', 'G', 'Q'], ['F', 'F', 'K', 'N'], ['F', 'F', 'N', 'Q'], ['F', 'G', 'G', 'G', 'M', 'S'], ['F', 'G', 'G', 'S', 'S', 'T'], ['F', 'G', 'M', 'N', 'S'], ['F', 'I', 'I', 'Y'], ['F', 'I', 'L', 'Y'], ['F', 'K', 'S', 'S', 'S'], ['F', 'L', 'L', 'Y'], ['F', 'N', 'S', 'S', 'T'], ['F', 'Q', 'S', 'S', 'S'], ['G', 'G', 'G', 'G', 'G', 'G', 'G', 'H'], ['G', 'G', 'G', 'G', 'G', 'G', 'P', 'P'], ['G', 'G', 'G', 'G', 'G', 'H', 'N'], ['G', 'G', 'G', 'G', 'N', 'P', 'P'], ['G', 'G', 'G', 'H', 'M', 'P'], ['G', 'G', 'G', 'H', 'N', 'N'], ['G', 'G', 'G', 'T', 'T', 'Y'], ['G', 'G', 'H', 'P', 'S', 'T'], ['G', 'G', 'H', 'S', 'V', 'V'], ['G', 'G', 'H', 'V', 'W'], ['G', 'G', 'I', 'I', 'P', 'V'], ['G', 'G', 'I', 'L', 'P', 'V'], ['G', 'G', 'K', 'M', 'Y'], ['G', 'G', 'L', 'L', 'P', 'V'], ['G', 'G', 'M', 'P', 'P', 'P'], ['G', 'G', 'M', 'Q', 'Y'], ['G', 'G', 'N', 'N', 'P', 'P'], ['G', 'H', 'I', 'K', 'T'], ['G', 'H', 'I', 'Q', 'T'], ['G', 'H', 'K', 'L', 'T'], ['G', 'H', 'L', 'Q', 'T'], ['G', 'H', 'M', 'N', 'P'], ['G', 'H', 'N', 'N', 'N'], ['G', 'H', 'R', 'S', 'V'], ['G', 'H', 'R', 'W'], ['G', 'I', 'I', 'P', 'R'], ['G', 'I', 'L', 'P', 'R'], ['G', 'K', 'S', 'T', 'Y'], ['G', 'L', 'L', 'P', 'R'], ['G', 'M', 'S', 'S', 'S', 'S'], ['G', 'N', 'T', 'T', 'Y'], ['G', 'P', 'P', 'P', 'S', 'T'], ['G', 'P', 'P', 'S', 'V', 'V'], ['G', 'P', 'P', 'V', 'W'], ['G', 'Q', 'S', 'T', 'Y'], ['H', 'H', 'M', 'M'], ['H', 'H', 'V', 'Y'], ['H', 'K', 'P', 'S', 'S'], ['H', 'N', 'P', 'S', 'T'], ['H', 'N', 'S', 'V', 'V'], ['H', 'N', 'V', 'W'], ['H', 'P', 'Q', 'S', 'S'], ['H', 'R', 'R', 'S'], ['I', 'I', 'N', 'P', 'V'], ['I', 'K', 'P', 'P', 'T'], ['I', 'K', 'P', 'V', 'V'], ['I', 'L', 'N', 'P', 'V'], ['I', 'P', 'P', 'Q', 'T'], ['I', 'P', 'Q', 'V', 'V'], ['I', 'P', 'Y', 'Y'], ['K', 'L', 'P', 'P', 'T'], ['K', 'L', 'P', 'V', 'V'], ['K', 'M', 'N', 'Y'], ['L', 'L', 'N', 'P', 'V'], ['L', 'P', 'P', 'Q', 'T'], ['L', 'P', 'Q', 'V', 'V'], ['L', 'P', 'Y', 'Y'], ['M', 'N', 'P', 'P', 'P'], ['M', 'N', 'Q', 'Y'], ['N', 'N', 'N', 'P', 'P'], ['P', 'P', 'R', 'S', 'V'], ['P', 'P', 'R', 'W'], ['S', 'S', 'S', 'S', 'S', 'T']]\n"
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "def branch(arr):\n",
    "    global arrsorted, score\n",
    "    alpha = [\"A\",\"C\",\"D\",\"E\",\"F\",\"G\",\"H\",\"I\",\"K\",\"L\",\"M\",\"N\",\"P\",\"Q\",\"R\",\"S\",\"T\",\"V\",\"W\",\"Y\"]\n",
    "    for i in range(len(alpha)):\n",
    "        if alpha[i] >= arr[len(arr)-1]:\n",
    "            summ = 0 \n",
    "            for j in range(len(arr)):\n",
    "                summ += Daltons[arr[j]]\n",
    "            if Daltons[alpha[i]] + summ == score:\n",
    "                arrsorted.append(arr+[alpha[i]])\n",
    "            elif Daltons[alpha[i]] + summ < score:\n",
    "                branch(arr+[alpha[i]])\n",
    "    \n",
    "def combo(arr):\n",
    "    score = Daltons[\"Y\"]+Daltons[\"M\"]+Daltons[\"K\"]+Daltons[\"N\"]\n",
    "    alpha = [\"A\",\"C\",\"D\",\"E\",\"F\",\"G\",\"H\",\"I\",\"K\",\"L\",\"M\",\"N\",\"P\",\"Q\",\"R\",\"S\",\"T\",\"V\",\"W\",\"Y\"]\n",
    "    for i in range(len(arr)):\n",
    "        for j in range(len(alpha)):\n",
    "            if alpha[j] >= arr[i][0]:\n",
    "                branch(arr[i]+[alpha[j]])\n",
    "                \n",
    "\n",
    "global arrsorted,score\n",
    "arrsorted = []\n",
    "\n",
    "Daltons = { 'A':  71, 'C': 103, 'D': 115, 'E': 129,\n",
    "           'F': 147, 'G':  57, 'H': 137, 'I': 113,\n",
    "           'K': 128, 'L': 113, 'M': 131, 'N': 114,\n",
    "           'P':  97, 'Q': 128, 'R': 156, 'S':  87,\n",
    "           'T': 101, 'V':  99, 'W': 186, 'Y': 163 }\n",
    "score = Daltons[\"Y\"]+Daltons[\"M\"]+Daltons[\"K\"]+Daltons[\"N\"]\n",
    "print score\n",
    "prod = []\n",
    "prod += itertools.product(Daltons,repeat=1)\n",
    "for i in range(len(prod)):\n",
    "    prod[i] = list(prod[i])\n",
    "print prod\n",
    "#prodsorted = []\n",
    "\n",
    "combo(prod)\n",
    "print len(arrsorted)\n",
    "print sorted(arrsorted)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 1,
    "section": "answer"
   },
   "source": [
    "[['A', 'A', 'A', 'A', 'D', 'H'], ['A', 'A', 'A', 'C', 'G', 'Y'], ['A', 'A', 'A', 'E', 'G', 'H'], ['A', 'A', 'A', 'E', 'P', 'P'], ['A', 'A', 'A', 'H', 'S', 'V'], ['A', 'A', 'A', 'H', 'W'], ['A', 'A', 'A', 'I', 'I', 'P'], ['A', 'A', 'A', 'I', 'L', 'P'], ['A', 'A', 'A', 'L', 'L', 'P'], ['A', 'A', 'C', 'C', 'G', 'M'], ['A', 'A', 'C', 'C', 'S', 'T'], ['A', 'A', 'C', 'F', 'G', 'S'], ['A', 'A', 'C', 'G', 'H', 'P'], ['A', 'A', 'C', 'K', 'Y'], ['A', 'A', 'C', 'P', 'P', 'P'], ['A', 'A', 'C', 'Q', 'Y'], ['A', 'A', 'E', 'H', 'K'], ['A', 'A', 'E', 'H', 'Q'], ['A', 'A', 'G', 'H', 'I', 'S'], ['A', 'A', 'G', 'H', 'L', 'S'], ['A', 'A', 'G', 'H', 'T', 'V'], ['A', 'A', 'G', 'S', 'S', 'Y'], ['A', 'A', 'H', 'R', 'T'], ['A', 'A', 'I', 'P', 'P', 'S'], ['A', 'A', 'L', 'P', 'P', 'S'], ['A', 'A', 'P', 'P', 'T', 'V'], ['A', 'A', 'P', 'V', 'V', 'V'], ['A', 'C', 'C', 'C', 'G', 'V'], ['A', 'C', 'C', 'C', 'R'], ['A', 'C', 'C', 'D', 'G', 'S'], ['A', 'C', 'C', 'G', 'T', 'T'], ['A', 'C', 'C', 'K', 'M'], ['A', 'C', 'C', 'M', 'Q'], ['A', 'C', 'F', 'G', 'G', 'T'], ['A', 'C', 'F', 'K', 'S'], ['A', 'C', 'F', 'N', 'T'], ['A', 'C', 'F', 'Q', 'S'], ['A', 'C', 'G', 'M', 'S', 'S'], ['A', 'C', 'H', 'K', 'P'], ['A', 'C', 'H', 'P', 'Q'], ['A', 'C', 'S', 'S', 'S', 'T'], ['A', 'D', 'G', 'G', 'H', 'V'], ['A', 'D', 'G', 'H', 'R'], ['A', 'D', 'G', 'P', 'P', 'V'], ['A', 'D', 'H', 'N', 'V'], ['A', 'D', 'P', 'P', 'R'], ['A', 'F', 'F', 'G', 'G', 'G'], ['A', 'F', 'F', 'G', 'N'], ['A', 'F', 'G', 'S', 'S', 'S'], ['A', 'G', 'G', 'G', 'M', 'Y'], ['A', 'G', 'G', 'H', 'I', 'T'], ['A', 'G', 'G', 'H', 'L', 'T'], ['A', 'G', 'G', 'S', 'T', 'Y'], ['A', 'G', 'H', 'P', 'S', 'S'], ['A', 'G', 'I', 'P', 'P', 'T'], ['A', 'G', 'I', 'P', 'V', 'V'], ['A', 'G', 'L', 'P', 'P', 'T'], ['A', 'G', 'L', 'P', 'V', 'V'], ['A', 'G', 'M', 'N', 'Y'], ['A', 'H', 'I', 'K', 'S'], ['A', 'H', 'I', 'N', 'T'], ['A', 'H', 'I', 'Q', 'S'], ['A', 'H', 'K', 'L', 'S'], ['A', 'H', 'K', 'T', 'V'], ['A', 'H', 'L', 'N', 'T'], ['A', 'H', 'L', 'Q', 'S'], ['A', 'H', 'Q', 'T', 'V'], ['A', 'I', 'P', 'R', 'V'], ['A', 'K', 'S', 'S', 'Y'], ['A', 'L', 'P', 'R', 'V'], ['A', 'N', 'S', 'T', 'Y'], ['A', 'P', 'P', 'P', 'S', 'S'], ['A', 'Q', 'S', 'S', 'Y'], ['C', 'C', 'C', 'G', 'G', 'I'], ['C', 'C', 'C', 'G', 'G', 'L'], ['C', 'C', 'C', 'I', 'N'], ['C', 'C', 'C', 'K', 'V'], ['C', 'C', 'C', 'L', 'N'], ['C', 'C', 'C', 'Q', 'V'], ['C', 'C', 'D', 'G', 'G', 'T'], ['C', 'C', 'D', 'K', 'S'], ['C', 'C', 'D', 'N', 'T'], ['C', 'C', 'D', 'Q', 'S'], ['C', 'C', 'E', 'G', 'G', 'S'], ['C', 'C', 'E', 'N', 'S'], ['C', 'C', 'G', 'S', 'S', 'V'], ['C', 'C', 'G', 'S', 'W'], ['C', 'C', 'K', 'T', 'T'], ['C', 'C', 'Q', 'T', 'T'], ['C', 'C', 'R', 'S', 'S'], ['C', 'D', 'F', 'G', 'G', 'G'], ['C', 'D', 'F', 'G', 'N'], ['C', 'D', 'G', 'S', 'S', 'S'], ['C', 'F', 'G', 'K', 'T'], ['C', 'F', 'G', 'Q', 'T'], ['C', 'G', 'G', 'G', 'M', 'M'], ['C', 'G', 'G', 'G', 'V', 'Y'], ['C', 'G', 'G', 'M', 'S', 'T'], ['C', 'G', 'G', 'R', 'Y'], ['C', 'G', 'M', 'M', 'N'], ['C', 'G', 'N', 'V', 'Y'], ['C', 'G', 'S', 'S', 'T', 'T'], ['C', 'K', 'M', 'S', 'S'], ['C', 'M', 'N', 'S', 'T'], ['C', 'M', 'Q', 'S', 'S'], ['C', 'N', 'R', 'Y'], ['D', 'E', 'E', 'Y'], ['D', 'F', 'H', 'H'], ['D', 'G', 'G', 'G', 'H', 'I'], ['D', 'G', 'G', 'G', 'H', 'L'], ['D', 'G', 'G', 'G', 'S', 'Y'], ['D', 'G', 'G', 'I', 'P', 'P'], ['D', 'G', 'G', 'L', 'P', 'P'], ['D', 'G', 'H', 'I', 'N'], ['D', 'G', 'H', 'K', 'V'], ['D', 'G', 'H', 'L', 'N'], ['D', 'G', 'H', 'Q', 'V'], ['D', 'G', 'N', 'S', 'Y'], ['D', 'H', 'K', 'R'], ['D', 'H', 'Q', 'R'], ['D', 'I', 'N', 'P', 'P'], ['D', 'K', 'P', 'P', 'V'], ['D', 'L', 'N', 'P', 'P'], ['D', 'P', 'P', 'Q', 'V'], ['E', 'E', 'F', 'M'], ['E', 'F', 'F', 'I'], ['E', 'F', 'F', 'L'], ['E', 'F', 'P', 'Y'], ['E', 'G', 'G', 'G', 'H', 'V'], ['E', 'G', 'G', 'H', 'R'], ['E', 'G', 'G', 'P', 'P', 'V'], ['E', 'G', 'H', 'N', 'V'], ['E', 'G', 'P', 'P', 'R'], ['E', 'H', 'N', 'R'], ['E', 'I', 'M', 'Y'], ['E', 'L', 'M', 'Y'], ['E', 'N', 'P', 'P', 'V'], ['F', 'F', 'G', 'G', 'K'], ['F', 'F', 'G', 'G', 'Q'], ['F', 'F', 'K', 'N'], ['F', 'F', 'N', 'Q'], ['F', 'G', 'G', 'G', 'M', 'S'], ['F', 'G', 'G', 'S', 'S', 'T'], ['F', 'G', 'M', 'N', 'S'], ['F', 'I', 'I', 'Y'], ['F', 'I', 'L', 'Y'], ['F', 'K', 'S', 'S', 'S'], ['F', 'L', 'L', 'Y'], ['F', 'N', 'S', 'S', 'T'], ['F', 'Q', 'S', 'S', 'S'], ['G', 'G', 'G', 'G', 'G', 'G', 'G', 'H'], ['G', 'G', 'G', 'G', 'G', 'G', 'P', 'P'], ['G', 'G', 'G', 'G', 'G', 'H', 'N'], ['G', 'G', 'G', 'G', 'N', 'P', 'P'], ['G', 'G', 'G', 'H', 'M', 'P'], ['G', 'G', 'G', 'H', 'N', 'N'], ['G', 'G', 'G', 'T', 'T', 'Y'], ['G', 'G', 'H', 'P', 'S', 'T'], ['G', 'G', 'H', 'S', 'V', 'V'], ['G', 'G', 'H', 'V', 'W'], ['G', 'G', 'I', 'I', 'P', 'V'], ['G', 'G', 'I', 'L', 'P', 'V'], ['G', 'G', 'K', 'M', 'Y'], ['G', 'G', 'L', 'L', 'P', 'V'], ['G', 'G', 'M', 'P', 'P', 'P'], ['G', 'G', 'M', 'Q', 'Y'], ['G', 'G', 'N', 'N', 'P', 'P'], ['G', 'H', 'I', 'K', 'T'], ['G', 'H', 'I', 'Q', 'T'], ['G', 'H', 'K', 'L', 'T'], ['G', 'H', 'L', 'Q', 'T'], ['G', 'H', 'M', 'N', 'P'], ['G', 'H', 'N', 'N', 'N'], ['G', 'H', 'R', 'S', 'V'], ['G', 'H', 'R', 'W'], ['G', 'I', 'I', 'P', 'R'], ['G', 'I', 'L', 'P', 'R'], ['G', 'K', 'S', 'T', 'Y'], ['G', 'L', 'L', 'P', 'R'], ['G', 'M', 'S', 'S', 'S', 'S'], ['G', 'N', 'T', 'T', 'Y'], ['G', 'P', 'P', 'P', 'S', 'T'], ['G', 'P', 'P', 'S', 'V', 'V'], ['G', 'P', 'P', 'V', 'W'], ['G', 'Q', 'S', 'T', 'Y'], ['H', 'H', 'M', 'M'], ['H', 'H', 'V', 'Y'], ['H', 'K', 'P', 'S', 'S'], ['H', 'N', 'P', 'S', 'T'], ['H', 'N', 'S', 'V', 'V'], ['H', 'N', 'V', 'W'], ['H', 'P', 'Q', 'S', 'S'], ['H', 'R', 'R', 'S'], ['I', 'I', 'N', 'P', 'V'], ['I', 'K', 'P', 'P', 'T'], ['I', 'K', 'P', 'V', 'V'], ['I', 'L', 'N', 'P', 'V'], ['I', 'P', 'P', 'Q', 'T'], ['I', 'P', 'Q', 'V', 'V'], ['I', 'P', 'Y', 'Y'], ['K', 'L', 'P', 'P', 'T'], ['K', 'L', 'P', 'V', 'V'], ['K', 'M', 'N', 'Y'], ['L', 'L', 'N', 'P', 'V'], ['L', 'P', 'P', 'Q', 'T'], ['L', 'P', 'Q', 'V', 'V'], ['L', 'P', 'Y', 'Y'], ['M', 'N', 'P', 'P', 'P'], ['M', 'N', 'Q', 'Y'], ['N', 'N', 'N', 'P', 'P'], ['P', 'P', 'R', 'S', 'V'], ['P', 'P', 'R', 'W'], ['S', 'S', 'S', 'S', 'S', 'T']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 2,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #2:** The primary enzyme complex involved in prokaryotic DNA replication is the **DNA polymerase III holoenzyme**. It is comprised of 9 proteins with 6 distinct types. The *HolC* gene encodes the &chi; component of this polymerase. The peptide sequence for *HolC* in [*Salmonella Typhimurium*](http://csbio.unc.edu/mcmillan/Comp555S18/data/SalmonellaTyphimurium.fa) is: \n",
    "\n",
    "         HolC = \"MTTHLNPPQEAPYMKNATFYLLDNDTTVNGLSAVEQLVCEIAAERWRAGK\" + \\\n",
    "                \"RVLIACEDEKQAIRLDEALWARPAESFVPHNLAGEGPRGGAPVEIAWPQK\" + \\\n",
    "                \"RNSSPRDILISLRTSFADFATAFTEVVDFVPYEETLKQLARERYKAYRVA\" + \\\n",
    "                \"GFNLNTATWK\"\n",
    "       \n",
    "How many possible DNA-sequence encodings are possible for this protein?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M ['AUG']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "H ['CAC', 'CAU']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "N ['AAC', 'AAU']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "Q ['CAG', 'CAA']\n",
      "E ['GAG', 'GAA']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "Y ['UAU', 'UAC']\n",
      "M ['AUG']\n",
      "K ['AAG', 'AAA']\n",
      "N ['AAC', 'AAU']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "F ['UUU', 'UUC']\n",
      "Y ['UAU', 'UAC']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "D ['GAU', 'GAC']\n",
      "N ['AAC', 'AAU']\n",
      "D ['GAU', 'GAC']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "N ['AAC', 'AAU']\n",
      "G ['GGU', 'GGG', 'GGA', 'GGC']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "S ['AGC', 'AGU', 'UCU', 'UCG', 'UCC', 'UCA']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "E ['GAG', 'GAA']\n",
      "Q ['CAG', 'CAA']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "C ['UGU', 'UGC']\n",
      "E ['GAG', 'GAA']\n",
      "I ['AUA', 'AUC', 'AUU']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "E ['GAG', 'GAA']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "W ['UGG']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "G ['GGU', 'GGG', 'GGA', 'GGC']\n",
      "K ['AAG', 'AAA']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "I ['AUA', 'AUC', 'AUU']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "C ['UGU', 'UGC']\n",
      "E ['GAG', 'GAA']\n",
      "D ['GAU', 'GAC']\n",
      "E ['GAG', 'GAA']\n",
      "K ['AAG', 'AAA']\n",
      "Q ['CAG', 'CAA']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "I ['AUA', 'AUC', 'AUU']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "D ['GAU', 'GAC']\n",
      "E ['GAG', 'GAA']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "W ['UGG']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "E ['GAG', 'GAA']\n",
      "S ['AGC', 'AGU', 'UCU', 'UCG', 'UCC', 'UCA']\n",
      "F ['UUU', 'UUC']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "H ['CAC', 'CAU']\n",
      "N ['AAC', 'AAU']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "G ['GGU', 'GGG', 'GGA', 'GGC']\n",
      "E ['GAG', 'GAA']\n",
      "G ['GGU', 'GGG', 'GGA', 'GGC']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "G ['GGU', 'GGG', 'GGA', 'GGC']\n",
      "G ['GGU', 'GGG', 'GGA', 'GGC']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "E ['GAG', 'GAA']\n",
      "I ['AUA', 'AUC', 'AUU']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "W ['UGG']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "Q ['CAG', 'CAA']\n",
      "K ['AAG', 'AAA']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "N ['AAC', 'AAU']\n",
      "S ['AGC', 'AGU', 'UCU', 'UCG', 'UCC', 'UCA']\n",
      "S ['AGC', 'AGU', 'UCU', 'UCG', 'UCC', 'UCA']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "D ['GAU', 'GAC']\n",
      "I ['AUA', 'AUC', 'AUU']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "I ['AUA', 'AUC', 'AUU']\n",
      "S ['AGC', 'AGU', 'UCU', 'UCG', 'UCC', 'UCA']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "S ['AGC', 'AGU', 'UCU', 'UCG', 'UCC', 'UCA']\n",
      "F ['UUU', 'UUC']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "D ['GAU', 'GAC']\n",
      "F ['UUU', 'UUC']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "F ['UUU', 'UUC']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "E ['GAG', 'GAA']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "D ['GAU', 'GAC']\n",
      "F ['UUU', 'UUC']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "P ['CCC', 'CCA', 'CCU', 'CCG']\n",
      "Y ['UAU', 'UAC']\n",
      "E ['GAG', 'GAA']\n",
      "E ['GAG', 'GAA']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "K ['AAG', 'AAA']\n",
      "Q ['CAG', 'CAA']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "E ['GAG', 'GAA']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "Y ['UAU', 'UAC']\n",
      "K ['AAG', 'AAA']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "Y ['UAU', 'UAC']\n",
      "R ['AGG', 'AGA', 'CGA', 'CGG', 'CGU', 'CGC']\n",
      "V ['GUU', 'GUC', 'GUG', 'GUA']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "G ['GGU', 'GGG', 'GGA', 'GGC']\n",
      "F ['UUU', 'UUC']\n",
      "N ['AAC', 'AAU']\n",
      "L ['CUU', 'CUG', 'CUA', 'CUC', 'UUG', 'UUA']\n",
      "N ['AAC', 'AAU']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "A ['GCA', 'GCG', 'GCC', 'GCU']\n",
      "T ['ACC', 'ACA', 'ACU', 'ACG']\n",
      "W ['UGG']\n",
      "K ['AAG', 'AAA']\n",
      "277854082212010956505804149831122269065829934762767371301132511301926141288251392 possible sequences\n"
     ]
    }
   ],
   "source": [
    "HolC = \"MTTHLNPPQEAPYMKNATFYLLDNDTTVNGLSAVEQLVCEIAAERWRAGK\" + \\\n",
    "            \"RVLIACEDEKQAIRLDEALWARPAESFVPHNLAGEGPRGGAPVEIAWPQK\" + \\\n",
    "            \"RNSSPRDILISLRTSFADFATAFTEVVDFVPYEETLKQLARERYKAYRVA\" + \\\n",
    "            \"GFNLNTATWK\"\n",
    "\"\"\"\n",
    "seq = []\n",
    "for i in range(0,len(HolC)-8):\n",
    "    cur = HolC[i:i+9]\n",
    "    alpha = []\n",
    "    for j in range(9):\n",
    "        if cur[j] not in alpha:\n",
    "            x = cur[j]\n",
    "            alpha.append(x)\n",
    "    if len(alpha)==6 and cur not in seq:\n",
    "        seq.append(cur)\n",
    "print len(seq)\n",
    "print seq\n",
    "\"\"\"\n",
    "codon = { # Maps an RNA triplet of nucelotides to a 1-letter Amino Acid Abbrevation\n",
    " \"AAA\": 'K', \"AAG\": 'K', \"AAC\": 'N', \"AAU\": 'N',\n",
    " \"AGA\": 'R', \"AGG\": 'R', \"AGC\": 'S', \"AGU\": 'S',\n",
    " \"ACA\": 'T', \"ACG\": 'T', \"ACC\": 'T', \"ACU\": 'T',\n",
    " \"AUA\": 'I', \"AUG\": 'M', \"AUC\": 'I', \"AUU\": 'I',\n",
    " \"GAA\": 'E', \"GAG\": 'E', \"GAC\": 'D', \"GAU\": 'D',\n",
    " \"GGA\": 'G', \"GGG\": 'G', \"GGC\": 'G', \"GGU\": 'G',\n",
    " \"GCA\": 'A', \"GCG\": 'A', \"GCC\": 'A', \"GCU\": 'A',\n",
    " \"GUA\": 'V', \"GUG\": 'V', \"GUC\": 'V', \"GUU\": 'V',\n",
    " \"CAA\": 'Q', \"CAG\": 'Q', \"CAC\": 'H', \"CAU\": 'H',\n",
    " \"CGA\": 'R', \"CGG\": 'R', \"CGC\": 'R', \"CGU\": 'R',\n",
    " \"CCA\": 'P', \"CCG\": 'P', \"CCC\": 'P', \"CCU\": 'P',\n",
    " \"CUA\": 'L', \"CUG\": 'L', \"CUC\": 'L', \"CUU\": 'L',\n",
    " \"UAA\": '*', \"UAG\": '*', \"UAC\": 'Y', \"UAU\": 'Y',\n",
    " \"UGA\": '*', \"UGG\": 'W', \"UGC\": 'C', \"UGU\": 'C',\n",
    " \"UCA\": 'S', \"UCG\": 'S', \"UCC\": 'S', \"UCU\": 'S',\n",
    " \"UUA\": 'L', \"UUG\": 'L', \"UUC\": 'F', \"UUU\": 'F'\n",
    "}\n",
    "AminoAcid = { # Maps 1-letter Amino Acid Abbrevations to their full name\n",
    "    'A': 'Alanine', 'C': 'Cysteine', 'D': 'Aspartic acid', 'E': 'Glutamic acid', 'F': 'Phenylalanine',\n",
    "    'G': 'Glycine', 'H': 'Histidine', 'I': 'Isoleucine', 'K': 'Lysine', 'L': 'Leucine', 'M': 'Methionine',\n",
    "    'N': 'Asparagine', 'P': 'Proline', 'Q': 'Glutamine', 'R': 'Arginine', 'S': 'Serine',\n",
    "    'T': 'Theronine', 'V': 'Valine', 'W': 'Tryptophan', 'Y': 'Tyrosine', '*': 'STOP'\n",
    "}\n",
    "\n",
    "AminoAbbrv = { # Maps 1-letter Amino Acid Abbrevations to 3-letter Abbrevations\n",
    "    'A': 'Ala', 'C': 'Cys', 'D': 'Asp', 'E': 'Glu', 'F': 'Phe', 'G': 'Gly', 'H': 'His', 'I': 'Ile',\n",
    "    'K': 'Lys', 'L': 'Leu', 'M': 'Met', 'N': 'Asn', 'P': 'Pro', 'Q': 'Gln', 'R': 'Arg', 'S': 'Ser',\n",
    "    'T': 'Thr', 'V': 'Val', 'W': 'Trp', 'Y': 'Tyr', '*': 'STP'    \n",
    "}\n",
    "TrimerCodes = {}\n",
    "for key, code in codon.iteritems():\n",
    "    TrimerCodes[code] = TrimerCodes.get(code,[]) + [key]\n",
    "codes = 1\n",
    "for residue in HolC:\n",
    "    print residue, TrimerCodes[residue]\n",
    "    codes *= len(TrimerCodes[residue])\n",
    "print \"%d possible sequences\" % codes"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 2,
    "section": "answer"
   },
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 3,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #3:** Find the coding sequence for the *HolC* in [*Salmonella Typhimurium*](http://csbio.unc.edu/mcmillan/Comp555S18/data/SalmonellaTyphimurium.fa). Your answer should give the sequence's slice interval in the genome (i.e. seq[start:end], where end=start+3&times;160) from the start codon to just before the stop codon. Note the the gene sequence might appear either on the primary sequence or its reverse complement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "137082\n",
      "4857432\n",
      "4720351\n",
      "4719871\n",
      "   137082 TTCCAGGTTGCCGTATTCAGGTTAAAACCAGCCACGCGGTAGGCTTTGTAGCGTTCGCGCGCCAGTTGTTTCAAAGTTTCTTCGTAAGGGACGAAGTCTACCACTTCTGTGAAAGCGGTGGCAAAATCTGCAAAGCTTGTACGCAGGCTGATTAAAATATCGCGCGGGCTGCTGTTGCGTTTTTGCGGCCAGGCGATTTCAACCGGCGCGCCGCCGCGTGGGCCTTCGCCTGCCAGATTGTGCGGGACAAAACTTTCCGCCGGTCTTGCCCACAGCGCTTCATCCAGCCGAATGGCCTGCTTCTCATCTTCGCAGGCGATCAGCACGCGCTTGCCCGCGCGCCAACGTTCTGCGGCAATTTCACACACCAGTTGTTCAACGGCGCTTAAACCGTTGACGGTGGTGTCATTATCCAGAAGATAGAACGTCGCATTTTTCATATATGGGGCTTCTTGTGGTGGATTTAAATGCGTTGTCATG UUCCAGGUUGCCGUAUUCAGGUUAAAACCAGCCACGCGGUAGGCUUUGUAGCGUUCGCGCGCCAGUUGUUUCAAAGUUUCUUCGUAAGGGACGAAGUCUACCACUUCUGUGAAAGCGGUGGCAAAAUCUGCAAAGCUUGUACGCAGGCUGAUUAAAAUAUCGCGCGGGCUGCUGUUGCGUUUUUGCGGCCAGGCGAUUUCAACCGGCGCGCCGCCGCGUGGGCCUUCGCCUGCCAGAUUGUGCGGGACAAAACUUUCCGCCGGUCUUGCCCACAGCGCUUCAUCCAGCCGAAUGGCCUGCUUCUCAUCUUCGCAGGCGAUCAGCACGCGCUUGCCCGCGCGCCAACGUUCUGCGGCAAUUUCACACACCAGUUGUUCAACGGCGCUUAAACCGUUGACGGUGGUGUCAUUAUCCAGAAGAUAGAACGUCGCAUUUUUCAUAUAUGGGGCUUCUUGUGGUGGAUUUAAAUGCGUUGUCAUG\n"
     ]
    }
   ],
   "source": [
    "def revComp(dnaSeq):\n",
    "    return ''.join([{'A':'T','C':'G','G':'C','T':'A'}[base] for base in reversed(dnaSeq)])\n",
    "\n",
    "def proteinTranslation(dnaSeq):\n",
    "    rnaSeq = dnaSeq.replace(\"T\", \"U\")\n",
    "    protSeq = ''\n",
    "    #for i in xrange(0,len(dnaSeq),3):\n",
    "        #if (i+3 > len(dnaSeq)):\n",
    "        #    break\n",
    "        #protSeq += AminoAbbrv[codon[rnaSeq[i:i+3]]]\n",
    "    return rnaSeq#, protSeq\n",
    "def loadFasta(filename):\n",
    "    \"\"\" Parses a classically formatted and possibly \n",
    "        compressed FASTA file into a list of headers \n",
    "        and fragment sequences for each sequence contained\"\"\"\n",
    "    if (filename.endswith(\".gz\")):\n",
    "        fp = gzip.open(filename, 'rb')\n",
    "    else:\n",
    "        fp = open(filename, 'rb')\n",
    "    # split at headers\n",
    "    data = fp.read().split('>')\n",
    "    fp.close()\n",
    "    # ignore whatever appears before the 1st header\n",
    "    data.pop(0)     \n",
    "    headers = []\n",
    "    sequences = []\n",
    "    for sequence in data:\n",
    "        lines = sequence.split('\\n')\n",
    "        headers.append(lines.pop(0))\n",
    "        # add an extra \"+\" to make string \"1-referenced\"\n",
    "        sequences.append('+' + ''.join(lines))\n",
    "    return (headers, sequences)\n",
    "\n",
    "header, seq = loadFasta(\"SalmonellaTyphimurium.fa\")\n",
    "seq = seq[0]\n",
    "seq = seq[1:len(seq)]\n",
    "print len(seq)\n",
    "revseq=revComp(seq)\n",
    "#seq = proteinTranslation(seq)\n",
    "#revseq=proteinTranslation(revseq)\n",
    "def AminoAcidSeqCount(genome, peptideSeq):   # Note: Only checks primary strand\n",
    "    readframe = []\n",
    "    anchor = peptideSeq[0]\n",
    "    for rnaSeq in TrimerCodes[anchor]:\n",
    "        dnaSeq = rnaSeq.replace('U','T')\n",
    "        start = 0\n",
    "        while True:\n",
    "            pos = genome.find(dnaSeq, start)\n",
    "            if (pos < 0):\n",
    "                break\n",
    "            i = 1\n",
    "            while (i < len(peptideSeq) and (pos+3*(i+1) <= len(genome))):\n",
    "                if (codon[genome[pos+3*i:pos+3*(i+1)].replace('T','U')] != peptideSeq[i]):\n",
    "                    break\n",
    "                i += 1\n",
    "            else:\n",
    "                readframe.append(pos)\n",
    "            start = pos + 1\n",
    "        return readframe\n",
    "\n",
    "#\n",
    "#Everything above here was basically taken from his code\n",
    "#This for loop below me goes through each letter of the peptide, finds where the peptide sequence occurs in our genome\n",
    "#and then keeps going until it gets throught he whole peptide\n",
    "peptide = HolC\n",
    "\n",
    "    ORFs = AminoAcidSeqCount(revseq,peptide[:i])\n",
    "    #print \"Searching for\", peptide[:i], \"Found it\", len(ORFs), \"times\" \n",
    "#print seq[137082:137262]\n",
    "#print seq[4719871:4720351]\n",
    "\n",
    "#what we found was that the peptide sequence didnt occur in the regular sequence, but the reverse complement\n",
    "#therefore the position that was returned was actually on the reversed sequence, so we can subtract it from the\n",
    "#end of ours and find where the sequence should start\n",
    "for pos in ORFs:\n",
    "    start = pos\n",
    "    print pos\n",
    "    print len(seq)\n",
    "    print len(seq)-start+1\n",
    "    print len(seq)-start-480+1\n",
    "    end = pos + (3*len(peptide))\n",
    "    print \"%9d %s %s\" % (start, seq[len(seq)-start-480+1:len(seq)-start+1], proteinTranslation(seq[len(seq)-start-480+1:len(seq)-start+1]))\n",
    "\n",
    "\"\"\"\n",
    "What we have done here \n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 3,
    "section": "answer"
   },
   "source": [
    "What we discovered is that the coding sequence for HolC was contained in the reverse complement of the Salmonella Typhimurium genome. Therefore we have to take the position we were given and subtract it from the total length of the genome. This will give us the position for the end of the sequence in our normal genome. We then subtract 3 * peptide length to get the start. \n",
    "From index [4719871:4720351]\n",
    "TTCCAGGTTGCCGTATTCAGGTTAAAACCAGCCACGCGGTAGGCTTTGTAGCGTTCGCGCGCCAGTTGTTTCAAAGTTTCTTCGTAAGGGACGAAGTCTACCACTTCTGTGAAAGCGGTGGCAAAATCTGCAAAGCTTGTACGCAGGCTGATTAAAATATCGCGCGGGCTGCTGTTGCGTTTTTGCGGCCAGGCGATTTCAACCGGCGCGCCGCCGCGTGGGCCTTCGCCTGCCAGATTGTGCGGGACAAAACTTTCCGCCGGTCTTGCCCACAGCGCTTCATCCAGCCGAATGGCCTGCTTCTCATCTTCGCAGGCGATCAGCACGCGCTTGCCCGCGCGCCAACGTTCTGCGGCAATTTCACACACCAGTTGTTCAACGGCGCTTAAACCGTTGACGGTGGTGTCATTATCCAGAAGATAGAACGTCGCATTTTTCATATATGGGGCTTCTTGTGGTGGATTTAAATGCGTTGTCATG UUCCAGGUUGCCGUAUUCAGGUUAAAACCAGCCACGCGGUAGGCUUUGUAGCGUUCGCGCGCCAGUUGUUUCAAAGUUUCUUCGUAAGGGACGAAGUCUACCACUUCUGUGAAAGCGGUGGCAAAAUCUGCAAAGCUUGUACGCAGGCUGAUUAAAAUAUCGCGCGGGCUGCUGUUGCGUUUUUGCGGCCAGGCGAUUUCAACCGGCGCGCCGCCGCGUGGGCCUUCGCCUGCCAGAUUGUGCGGGACAAAACUUUCCGCCGGUCUUGCCCACAGCGCUUCAUCCAGCCGAAUGGCCUGCUUCUCAUCUUCGCAGGCGAUCAGCACGCGCUUGCCCGCGCGCCAACGUUCUGCGGCAAUUUCACACACCAGUUGUUCAACGGCGCUUAAACCGUUGACGGUGGUGUCAUUAUCCAGAAGAUAGAACGUCGCAUUUUUCAUAUAUGGGGCUUCUUGUGGUGGAUUUAAAUGCGUUGUCAUG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 4,
    "section": "problem"
   },
   "source": [
    "**Problem #4:** How many *distinct* subpeptides can the *HolC* protein from problems 3 and 4 be broken into? How many *distinct* molecular weights appear in its Theoretical Spectrum?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12704\n",
      "7909\n"
     ]
    }
   ],
   "source": [
    "#this is just the kmerfreq function that finds all combinations of the subsequences, it also removes duplicates\n",
    "def kmerFreq(k, sequence):\n",
    "    kmerCount = {}\n",
    "    for i in xrange(len(sequence)-k+1):\n",
    "        kmer = sequence[i:i+k]\n",
    "        kmerCount[kmer] = kmerCount.get(kmer,0)+1\n",
    "    return kmerCount\n",
    "summ = 0\n",
    "for i in range(len(HolC)):\n",
    "    summ += len(kmerFreq(i, HolC))\n",
    "print summ\n",
    "Daltons = { \n",
    "    'A':  71, 'C': 103, 'D': 115, 'E': 129,\n",
    "    'F': 147, 'G':  57, 'H': 137, 'I': 113,\n",
    "    'K': 128, 'L': 113, 'M': 131, 'N': 114,\n",
    "    'P':  97, 'Q': 128, 'R': 156, 'S':  87,\n",
    "    'T': 101, 'V':  99, 'W': 186, 'Y': 163 \n",
    "}\n",
    "\n",
    "#this was also from his code, it just calculates the number of possible molecular weights\n",
    "def TheoreticalSpectrum(peptide):\n",
    "    # Generate every possible fragment of a peptide\n",
    "    spectrum = set()\n",
    "    for fragLength in xrange(1,len(peptide)+1):\n",
    "        for start in xrange(0,len(peptide)-fragLength+1):\n",
    "            seq = peptide[start:start+fragLength]\n",
    "            spectrum.add(sum([Daltons[res] for res in seq]))\n",
    "    return sorted(spectrum)\n",
    "spectrum = TheoreticalSpectrum(HolC)\n",
    "print len(spectrum)\n",
    "\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 4,
    "section": "answer"
   },
   "source": [
    "Distinct subpeptides: 12704\n",
    "Distinct molecular weights: 7909"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 5,
    "section": "problem"
   },
   "source": [
    "**Problem #5: Programming Problem**\n",
    "\n",
    "In class, we discussed techniques for finding occurrences of a motif in a larger DNA string. However, because a DNA strand coding for a protein is often interspersed with introns (segments of DNA that do not code for proteins), we need a way to recognize a motif that has been chopped up into pieces along a chromosome.\n",
    "\n",
    "A subsequence of a string is a collection of symbols contained in order (though not necessarily contiguously) in the string (e.g., ACG is a subsequence of TATGCTAAGATC). The indices of a subsequence are the positions in the string at which the symbols of the subsequence appear. For example, the indices of ACG in TATGCTAAGATCG can be represented by 6 combinations of indices: (2, 5, 9), (2, 5, 13), (2, 12, 13), (7, 12, 13), (8, 12, 13), and (10, 12, 13). \n",
    "\n",
    "Write a program that takes two strings: a string `s` and a sequence, and returns a list of index tuples for which `s` appears as a subsequence in the sequence:\n",
    "\n",
    "\n",
    "     findSubsequences(s, sequence)\n",
    "\n",
    "\n",
    "Apply your algorithm to the following set of sequences:\n",
    "\n",
    "<pre>\n",
    "test =  \"ggtaattgcctcgcataacgcggtgtgaaaatggattgaagcccgggcggtggattctac\"+\\\n",
    "        \"tcaactttagccgatggagaaagccccgggatccgggctattaaaaagaagatcttttta\"+\\\n",
    "        \"tttagagatctgttctattgtgatctcttattaggatcgcgccaggctgtggataacccg\"+\\\n",
    "        \"gatcctgtaataaagatcaatgcgttggaaaggatcactagctgtgaatgatcggtgatc\"\n",
    "\n",
    "subseq=\"gactc\"\n",
    "</pre>\n",
    "\n",
    "Use the %time *magic* function to gauge the performance of your method to the one given."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "section": "submit"
   },
   "source": [
    "Click [here to submit](http://csbio.unc.edu/mcmillan/index.py?run=PS.upload) your completed problem set. **Make sure that you do not overwrite a previous problem set!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "number": 5,
    "section": "answer"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2202\n",
      "31244\n",
      "546534\n",
      "4834514\n",
      "CPU times: user 11.8 s, sys: 522 ms, total: 12.3 s\n",
      "Wall time: 12.2 s\n"
     ]
    }
   ],
   "source": [
    "# Enter your program here\n",
    "test =  \"ggtaattgcctcgcataacgcggtgtgaaaatggattgaagcccgggcggtggattctac\"+\\\n",
    "        \"tcaactttagccgatggagaaagccccgggatccgggctattaaaaagaagatcttttta\"+\\\n",
    "        \"tttagagatctgttctattgtgatctcttattaggatcgcgccaggctgtggataacccg\"+\\\n",
    "        \"gatcctgtaataaagatcaatgcgttggaaaggatcactagctgtgaatgatcggtgatc\"\n",
    "\n",
    "subseq=\"gactc\"\n",
    "\n",
    "letters = []\n",
    "#this function creates an array consisting of d + any values of e greater than the last thing in d\n",
    "def recurcombine(start, d, e):\n",
    "    newstart = start\n",
    "    temp = []\n",
    "    #we loop through the possible indexes of the letter\n",
    "    for i in range(start,len(e)):\n",
    "        #we check to see if the index is greater than the last number in d\n",
    "        if e[i] > d[len(d)-1]:\n",
    "            #we add the index of the letter to the array d and then add it to our temporary array\n",
    "            temp.append(d + [e[i]])\n",
    "        else:\n",
    "            #if it was smaller, we know that we can move our starting position up, because the following\n",
    "            #arrays will always have a last number > the previous start\n",
    "            newstart = i\n",
    "    return newstart, temp\n",
    "\n",
    "#main function\n",
    "def findSubsequences(subseq,test):\n",
    "    #this here determines all the locations of each letter in our subseq\n",
    "    #ex. if our test was agta, and our subsequence was ag, it would return [[0,3],[1]]\n",
    "    #the first subarray refers to the letter a and the second to the letter g\n",
    "    for i in range(len(subseq)):\n",
    "        letters.append([])\n",
    "        for j in range(len(test)):\n",
    "            if subseq[i]==test[j]:\n",
    "                letters[i].append(j)\n",
    "    #we grab the first letter array and want to turn it into an array\n",
    "    \"\"\"\n",
    "    c=letters[0]\n",
    "    for i in range(len(c)):\n",
    "        c[i]=[c[i]]\n",
    "    \"\"\"\n",
    "    c = list(itertools.product(letters[0],letters[1]))\n",
    "    #print c\n",
    "    i = 0\n",
    "    while i < (len(c)):\n",
    "        x = len(c[i])-1\n",
    "        if c[i][x] < c[i][x-1]:\n",
    "            c.pop(i)\n",
    "        else:\n",
    "            c[i] = list(c[i])\n",
    "            i=i+1\n",
    "    a = len(c[0])-1\n",
    "    c.sort(key=lambda x: x[a])\n",
    "    #print c\n",
    "    #c = list(itertools.product(c,letters[2]))\n",
    "    \n",
    "\n",
    "    #now what we do is go through the rest of the subsequence\n",
    "    for z in range(2,len(subseq)):\n",
    "        #this array is to hold our newly calculated subsequences when we add another letter\n",
    "        newcomb = []\n",
    "        #this will hold our current starting position\n",
    "        start = 0\n",
    "        #this is a counter\n",
    "        j = 0\n",
    "        print len(c)\n",
    "        #we loop through all of our current subsequences\n",
    "        while j < len(c):\n",
    "            start, temp = recurcombine(start, c[j], letters[z])\n",
    "            #this will adjust our start position each time and also add new subsequences to our array\n",
    "            for x in temp:\n",
    "                newcomb.append(x)\n",
    "            j += 1\n",
    "        #we then set c to our new set of subsequences (which have a new letter added to it)\n",
    "        c = newcomb\n",
    "        a = z\n",
    "        #we then have to sort our array by the last index \n",
    "        c.sort(key=lambda x: x[a])\n",
    "    print len(c)\n",
    "\n",
    "%time findSubsequences(subseq, test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
