{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "section": "header"
   },
   "source": [
    "<div style=\"text-align: center;\">\n",
    "<h1>The University of North Carolina at Chapel Hill</h1>\n",
    "<h1>Comp 555 BioAlgorithms - Spring 2019</h1>\n",
    "<h1 style=\"font-size: 250%;\">Final Exam</h1>\n",
    "<h1>Friday May 3, 2019</h1>\n",
    "<h2>8:00am-11:00am (180 mins)</h2>\n",
    "</div>\n",
    "<br>\n",
    "\n",
    "**Instructions:** Answer all questions ***using the cells*** provided. Feel free to add additional cells to your notebook, but there is no guarantee that they will be considered when grading. If any question seems unclear or appears to lack sufficient information, state a reasonable assumption in the cell provided for your answer and proceed. This test is open-book, open-note, and open-internet rosources. However, ***you cannot use the internet as a means for communication with anyone interactively***. This means no instant messaging, no social-media communication, no virtual whiteboards, and no shared web-documents are allowed. In other words, any attempt to communicate with another person, other than the instructors, during the test will be considered an honor-code violation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 1,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #1:** (10 points) Given the following DNA alignment scoring matrix:\n",
    "\n",
    "<table style=\"border-collapse: collapse; border: 1px solid black; background-color: white;\">\n",
    "    <tr style=\"background-color: white;\">\n",
    "        <th>&delta;</th>\n",
    "        <th style=\"border: 1px solid black;\">  A </th>\n",
    "        <th style=\"border: 1px solid black;\">  C </th>\n",
    "        <th style=\"border: 1px solid black;\">  G </th>\n",
    "        <th style=\"border: 1px solid black;\">  T </th>\n",
    "    </tr>\n",
    "    <tr style=\"background-color: white;\">\n",
    "        <th style=\"border: 1px solid black;\">   A  </th>\n",
    "        <td style=\"background-color: #f0fff0;\">  2 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -2 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -1 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -2 </td>\n",
    "    </tr>\n",
    "    <tr style=\"background-color: white;\">\n",
    "        <th style=\"border: 1px solid black;\">   C  </th>\n",
    "        <td style=\"background-color: #fff0f0;\"> -2 </td>\n",
    "        <td style=\"background-color: #f0fff0;\">  2 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -2 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -1 </td>\n",
    "    </tr>\n",
    "    <tr style=\"background-color: white;\">\n",
    "        <th style=\"border: 1px solid black;\">   G  </th>\n",
    "        <td style=\"background-color: #fff0f0;\"> -1 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -2 </td>\n",
    "        <td style=\"background-color: #f0fff0;\">  2 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -2 </td>\n",
    "    </tr>\n",
    "    <tr style=\"background-color: white;\">\n",
    "        <th style=\"border: 1px solid black;\">   T  </th>\n",
    "        <td style=\"background-color: #fff0f0;\"> -2 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -1 </td>\n",
    "        <td style=\"background-color: #fff0f0;\"> -2 </td>\n",
    "        <td style=\"background-color: #f0fff0;\">  2 </td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n",
    "and an *InDel* penalty of -3, what is the best *global alignment score*, and alignment for the following DNA sequences?\n",
    "\n",
    "    v = \"AACACAAACTGAGCCCCTACCCCAAATTAGAACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCCATTTTACCAATGCATA\"\n",
    "    w = \"AAAAACACAAACTGAGCCCCTACCCCAAATTAGGACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCTATTTTACCAATGT\"\n",
    "\n",
    "In the space provided below give a ***number for your global alignment score*** and ***show an alignment*** in the style of the Alignment() method given in class (Lecture 18, slide 27). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "s\n",
      " [[   0   -3   -6 ... -264 -267 -270]\n",
      " [  -3    2   -1 ... -259 -262 -265]\n",
      " [  -6   -1    4 ... -254 -257 -260]\n",
      " ...\n",
      " [-264 -259 -254 ...  146  151  153]\n",
      " [-267 -262 -257 ...  143  148  153]\n",
      " [-270 -265 -260 ...  140  145  150]]\n",
      "b\n",
      " [[0 2 2 ... 2 2 2]\n",
      " [1 3 2 ... 2 2 2]\n",
      " [1 1 3 ... 2 2 2]\n",
      " ...\n",
      " [1 1 1 ... 1 1 1]\n",
      " [1 1 1 ... 1 1 3]\n",
      " [1 1 1 ... 1 1 1]]\n",
      "Best score = 150\n",
      "v = AA___CACAAACTGAGCCCCTACCCCAAATTAGAACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCCATTTTACCAATGCATA\n",
      "w = AAAAACACAAACTGAGCCCCTACCCCAAATTAGGACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCTATTTTACCAATG__T_\n"
     ]
    }
   ],
   "source": [
    "import numpy\n",
    "v = \"AACACAAACTGAGCCCCTACCCCAAATTAGAACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCCATTTTACCAATGCATA\"\n",
    "w = \"AAAAACACAAACTGAGCCCCTACCCCAAATTAGGACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCTATTTTACCAATGT\"\n",
    "\n",
    "def Alignment(b,v,w,i,j):\n",
    "    if ((i == 0) and (j == 0)):\n",
    "        return ['','']\n",
    "    if (b[i,j] == 3):\n",
    "        result = Alignment(b,v,w,i-1,j-1)\n",
    "        result[0] += v[i-1]\n",
    "        result[1] += w[j-1]\n",
    "        return result\n",
    "    if (b[i,j] == 2):\n",
    "        result = Alignment(b,v,w,i,j-1)\n",
    "        result[0] += \"_\"\n",
    "        result[1] += w[j-1]\n",
    "        return result\n",
    "    if (b[i,j] == 1):\n",
    "        result = Alignment(b,v,w,i-1,j)\n",
    "        result[0] += v[i-1]\n",
    "        result[1] += \"_\"\n",
    "        return result\n",
    "\n",
    "def GlobalAlign(v, w, scorematrix, indel):\n",
    "    s = numpy.zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    b = numpy.zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    for i in range(0,len(v)+1):\n",
    "        for j in range(0,len(w)+1):\n",
    "            if (j == 0):\n",
    "                if (i > 0):\n",
    "                    s[i,j] = s[i-1,j] + indel\n",
    "                    b[i,j] = 1\n",
    "                continue\n",
    "            if (i == 0):\n",
    "                s[i,j] = s[i,j-1] + indel\n",
    "                b[i,j] = 2\n",
    "                continue\n",
    "            score = s[i-1,j-1] + scorematrix[v[i-1],w[j-1]]\n",
    "            vskip = s[i-1,j] + indel\n",
    "            wskip = s[i,j-1] + indel\n",
    "            s[i,j] = max(vskip, wskip, score)\n",
    "            if (s[i,j] == vskip):\n",
    "                b[i,j] = 1\n",
    "            elif (s[i,j] == wskip):\n",
    "                b[i,j] = 2\n",
    "            else:\n",
    "                b[i,j] = 3\n",
    "    return (s, b)\n",
    "\n",
    "match = {('A','A'):  2, ('A','C'): -2, ('A','G'): -1, ('A','T'): -2,\n",
    "         ('C','A'): -2, ('C','C'):  2, ('C','G'): -2, ('C','T'): -1,\n",
    "         ('G','A'): -1, ('G','C'): -2, ('G','G'):  2, ('G','T'): -2,\n",
    "         ('T','A'): -2, ('T','C'): -1, ('T','G'): -2, ('T','T'):  2}\n",
    "\n",
    "# v = \"TTCCGAGCGTTA\"\n",
    "# w = \"TTTCAGGTTA\"\n",
    "\n",
    "s, b = GlobalAlign(v,w,match,-3)\n",
    "print(\"s\\n\",s)\n",
    "print(\"b\\n\",b)\n",
    "print(\"Best score =\", s[-1,-1])\n",
    "align = Alignment(b,v,w,b.shape[0]-1,b.shape[1]-1)\n",
    "print(\"v =\", align[0])\n",
    "print(\"w =\", align[1])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 1,
    "section": "answer"
   },
   "source": [
    "Enter the global alignment score, and an alignment that achieves it.\n",
    "\n",
    "Best score = 150\n",
    "v = AA___CACAAACTGAGCCCCTACCCCAAATTAGAACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCCATTTTACCAATGCATA\n",
    "w = AAAAACACAAACTGAGCCCCTACCCCAAATTAGGACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCTATTTTACCAATG__T_\n",
    "\n",
    "\n",
    "s\n",
    " [[   0   -3   -6 ... -264 -267 -270]\n",
    " [  -3    2   -1 ... -259 -262 -265]\n",
    " [  -6   -1    4 ... -254 -257 -260]\n",
    " ...\n",
    " [-264 -259 -254 ...  146  151  153]\n",
    " [-267 -262 -257 ...  143  148  153]\n",
    " [-270 -265 -260 ...  140  145  150]]\n",
    "b\n",
    " [[0 2 2 ... 2 2 2]\n",
    " [1 3 2 ... 2 2 2]\n",
    " [1 1 3 ... 2 2 2]\n",
    " ...\n",
    " [1 1 1 ... 1 1 1]\n",
    " [1 1 1 ... 1 1 3]\n",
    " [1 1 1 ... 1 1 1]]\n",
    "\n",
    "\n",
    "Code used:\n",
    "import numpy\n",
    "v = \"AACACAAACTGAGCCCCTACCCCAAATTAGAACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCCATTTTACCAATGCATA\"\n",
    "w = \"AAAAACACAAACTGAGCCCCTACCCCAAATTAGGACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCTATTTTACCAATGT\"\n",
    "\n",
    "def Alignment(b,v,w,i,j):\n",
    "    if ((i == 0) and (j == 0)):\n",
    "        return ['','']\n",
    "    if (b[i,j] == 3):\n",
    "        result = Alignment(b,v,w,i-1,j-1)\n",
    "        result[0] += v[i-1]\n",
    "        result[1] += w[j-1]\n",
    "        return result\n",
    "    if (b[i,j] == 2):\n",
    "        result = Alignment(b,v,w,i,j-1)\n",
    "        result[0] += \"_\"\n",
    "        result[1] += w[j-1]\n",
    "        return result\n",
    "    if (b[i,j] == 1):\n",
    "        result = Alignment(b,v,w,i-1,j)\n",
    "        result[0] += v[i-1]\n",
    "        result[1] += \"_\"\n",
    "        return result\n",
    "\n",
    "def GlobalAlign(v, w, scorematrix, indel):\n",
    "    s = numpy.zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    b = numpy.zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    for i in range(0,len(v)+1):\n",
    "        for j in range(0,len(w)+1):\n",
    "            if (j == 0):\n",
    "                if (i > 0):\n",
    "                    s[i,j] = s[i-1,j] + indel\n",
    "                    b[i,j] = 1\n",
    "                continue\n",
    "            if (i == 0):\n",
    "                s[i,j] = s[i,j-1] + indel\n",
    "                b[i,j] = 2\n",
    "                continue\n",
    "            score = s[i-1,j-1] + scorematrix[v[i-1],w[j-1]]\n",
    "            vskip = s[i-1,j] + indel\n",
    "            wskip = s[i,j-1] + indel\n",
    "            s[i,j] = max(vskip, wskip, score)\n",
    "            if (s[i,j] == vskip):\n",
    "                b[i,j] = 1\n",
    "            elif (s[i,j] == wskip):\n",
    "                b[i,j] = 2\n",
    "            else:\n",
    "                b[i,j] = 3\n",
    "    return (s, b)\n",
    "\n",
    "match = {('A','A'):  2, ('A','C'): -2, ('A','G'): -1, ('A','T'): -2,\n",
    "         ('C','A'): -2, ('C','C'):  2, ('C','G'): -2, ('C','T'): -1,\n",
    "         ('G','A'): -1, ('G','C'): -2, ('G','G'):  2, ('G','T'): -2,\n",
    "         ('T','A'): -2, ('T','C'): -1, ('T','G'): -2, ('T','T'):  2}\n",
    "\n",
    "s, b = GlobalAlign(v,w,match,-3)\n",
    "print(\"s\\n\",s)\n",
    "print(\"b\\n\",b)\n",
    "print(\"Best score =\", s[-1,-1])\n",
    "align = Alignment(b,v,w,b.shape[0]-1,b.shape[1]-1)\n",
    "print(\"v =\", align[0])\n",
    "print(\"w =\", align[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 2,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #2:** (10 points) Find the longest common subsequence between the <b>v</b> and <b>w</b> sequences given in **Problem #1**. Use it to find the minimum edit distance between these two sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[  0   0   0 ...   0   0   0]\n",
      " [  0   2   2 ...   0   0   0]\n",
      " [  0   2   4 ...   0   0   0]\n",
      " ...\n",
      " [  0   2   2 ... 143 152 158]\n",
      " [  0   0   0 ... 136 145 154]\n",
      " [  0   2   2 ... 129 138 147]]\n",
      "\n",
      "[[0 0 0 ... 0 0 0]\n",
      " [0 3 3 ... 0 0 0]\n",
      " [0 3 3 ... 3 0 0]\n",
      " ...\n",
      " [0 3 3 ... 1 1 1]\n",
      " [0 0 3 ... 1 1 3]\n",
      " [0 3 3 ... 1 1 1]]\n",
      "v: 1 : 86 ACACAAACTGAGCCCCTACCCCAAATTAGAACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCCATTTTACCAATG\n",
      "w: 1 : 89 AAAACACAAACTGAGCCCCTACCCCAAATTAGGACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCTATTTTACCAATG\n",
      "score: 166\n",
      "[(86, 89), (85, 88), (84, 87), (83, 86), (82, 85), (81, 84), (80, 83), (79, 82), (78, 81), (77, 80), (76, 79), (75, 78), (74, 77), (73, 76), (72, 75), (71, 74), (70, 73), (69, 72), (68, 71), (67, 70), (66, 69), (65, 68), (64, 67), (63, 66), (62, 65), (61, 64), (60, 63), (59, 62), (58, 61), (57, 60), (56, 59), (55, 58), (54, 57), (53, 56), (52, 55), (51, 54), (50, 53), (49, 52), (48, 51), (47, 50), (46, 49), (45, 48), (44, 47), (43, 46), (42, 45), (41, 44), (40, 43), (39, 42), (38, 41), (37, 40), (36, 39), (35, 38), (34, 37), (33, 36), (32, 35), (31, 35), (31, 34), (30, 33), (29, 32), (28, 31), (27, 30), (26, 29), (25, 28), (24, 27), (23, 26), (22, 25), (21, 24), (20, 23), (19, 22), (18, 21), (17, 20), (16, 19), (15, 18), (14, 17), (13, 16), (12, 15), (11, 14), (10, 13), (9, 12), (8, 11), (7, 10), (6, 9), (5, 8), (4, 7), (3, 6), (2, 5), (2, 4), (2, 3), (2, 2), (1, 2), (1, 1)]\n"
     ]
    }
   ],
   "source": [
    "match = {('A','A'):  2, ('A','C'): -2, ('A','G'): -1, ('A','T'): -2,\n",
    "         ('C','A'): -2, ('C','C'):  2, ('C','G'): -2, ('C','T'): -1,\n",
    "         ('G','A'): -1, ('G','C'): -2, ('G','G'):  2, ('G','T'): -2,\n",
    "         ('T','A'): -2, ('T','C'): -1, ('T','G'): -2, ('T','T'):  2}\n",
    "\n",
    "def LocalAlign(v, w, scorematrix, indel):\n",
    "    s = numpy.zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    b = numpy.zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    for i in range(1,len(v)+1):\n",
    "        for j in range(1,len(w)+1):\n",
    "            if (j == 0):\n",
    "                if (i > 0):\n",
    "                    s[i,j] = max(s[i-1,j] + indel, 0)\n",
    "                    b[i,j] = 1\n",
    "                continue\n",
    "            if (i == 0):\n",
    "                s[i,j] = max(s[i,j-1] + indel, 0)\n",
    "                b[i,j] = 2\n",
    "                continue\n",
    "            score = s[i-1,j-1] + scorematrix[v[i-1],w[j-1]]\n",
    "            vskip = s[i-1,j] + indel\n",
    "            wskip = s[i,j-1] + indel\n",
    "            s[i,j] = max(vskip, wskip, score, 0)\n",
    "            if (s[i,j] == vskip):\n",
    "                b[i,j] = 1\n",
    "            elif (s[i,j] == wskip):\n",
    "                b[i,j] = 2\n",
    "            elif (s[i,j] == score):\n",
    "                b[i,j] = 3\n",
    "            else:\n",
    "                b[i,j] = 0\n",
    "    return (s, b)\n",
    "\n",
    "s, b = LocalAlign(v,w,match,-7)\n",
    "print(s)\n",
    "print()\n",
    "print(b)\n",
    "\n",
    "def locBackTrack(i, j, score):\n",
    "    # v is i and w is j\n",
    "    vstart = i\n",
    "    wstart = j\n",
    "    path = [(i, j)]\n",
    "    val = score[i, j]\n",
    "    x = i\n",
    "    y = j\n",
    "    while val > 0:\n",
    "        left = score[vstart-1, wstart]\n",
    "        diag = score[vstart-1, wstart-1]\n",
    "        top = score[vstart, wstart - 1]\n",
    "        val = max(left, diag, top)\n",
    "        if val == 0:\n",
    "            break\n",
    "        if val == left:\n",
    "            vstart -= 1\n",
    "        elif val == diag:\n",
    "            vstart -= 1\n",
    "            wstart -= 1\n",
    "        elif val == top:\n",
    "            wstart -= 1\n",
    "        path.append((vstart, wstart))\n",
    "    return [vstart, i, wstart, j], path\n",
    "    \n",
    "    \n",
    "# we will create a function which will search for the highest score and backtrack from there\n",
    "def LocalFinding(v, w, scores, backtrack):\n",
    "    maxVal = -10000\n",
    "    loc = (0, 0)\n",
    "    for i in range(scores.shape[0]):\n",
    "        for j in range(scores.shape[1]):\n",
    "            if scores[i, j] > maxVal:\n",
    "                maxVal = scores[i, j]\n",
    "                loc = (i, j)\n",
    "    \n",
    "    # backtrack from our located position\n",
    "#     align = LocalAlignCreate(backtrack, scores, v, w, loc[0], loc[1])\n",
    "    return loc\n",
    "\n",
    "s, b = LocalAlign(v, w, match, -3)\n",
    "loc = LocalFinding(v, w, s, b)\n",
    "ranges, path = locBackTrack(loc[0], loc[1], s)\n",
    "print (\"v:\", ranges[0], \":\", ranges[1], v[ranges[0]:ranges[1]])\n",
    "print (\"w:\", ranges[2], \":\", ranges[3], w[ranges[2]:ranges[3]])\n",
    "print (\"score:\", s[loc[0], loc[1]])\n",
    "print (path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "s\n",
      " [[ 0  0  0 ...  0  0  0]\n",
      " [ 0  1  1 ...  1  1  1]\n",
      " [ 0  1  2 ...  2  2  2]\n",
      " ...\n",
      " [ 0  1  2 ... 83 84 84]\n",
      " [ 0  1  2 ... 83 84 85]\n",
      " [ 0  1  2 ... 83 84 85]]\n",
      "b\n",
      " [[0 0 0 ... 0 0 0]\n",
      " [0 3 3 ... 2 2 2]\n",
      " [0 3 3 ... 2 2 2]\n",
      " ...\n",
      " [0 3 3 ... 1 1 2]\n",
      " [0 1 1 ... 3 1 3]\n",
      " [0 3 3 ... 1 1 1]]\n",
      "90 90\n",
      "89 90\n",
      "88 89\n",
      "87 89\n",
      "86 89\n",
      "85 88\n",
      "84 87\n",
      "83 86\n",
      "82 85\n",
      "81 84\n",
      "80 83\n",
      "79 82\n",
      "78 81\n",
      "77 80\n",
      "76 79\n",
      "75 78\n",
      "74 77\n",
      "74 76\n",
      "73 75\n",
      "72 75\n",
      "71 74\n",
      "70 73\n",
      "69 72\n",
      "68 71\n",
      "67 70\n",
      "66 69\n",
      "65 68\n",
      "64 67\n",
      "63 66\n",
      "62 65\n",
      "61 64\n",
      "60 63\n",
      "59 62\n",
      "58 61\n",
      "57 60\n",
      "56 59\n",
      "55 58\n",
      "54 57\n",
      "53 56\n",
      "52 55\n",
      "51 54\n",
      "50 53\n",
      "49 52\n",
      "48 51\n",
      "47 50\n",
      "46 49\n",
      "45 48\n",
      "44 47\n",
      "43 46\n",
      "42 45\n",
      "41 44\n",
      "40 43\n",
      "39 42\n",
      "38 41\n",
      "37 40\n",
      "36 39\n",
      "35 38\n",
      "34 37\n",
      "33 36\n",
      "32 35\n",
      "31 34\n",
      "31 33\n",
      "30 33\n",
      "29 32\n",
      "28 31\n",
      "27 30\n",
      "26 29\n",
      "25 28\n",
      "24 27\n",
      "23 26\n",
      "22 25\n",
      "21 24\n",
      "20 23\n",
      "19 22\n",
      "18 21\n",
      "17 20\n",
      "16 19\n",
      "15 18\n",
      "14 17\n",
      "13 16\n",
      "12 15\n",
      "11 14\n",
      "10 13\n",
      "9 12\n",
      "8 11\n",
      "7 10\n",
      "6 9\n",
      "5 8\n",
      "4 7\n",
      "3 6\n",
      "2 5\n",
      "1 4\n",
      "0 3\n",
      "AACACAAACTGAGCCCCTACCCCAAATTAGACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCATTTTACCAATGT\n"
     ]
    }
   ],
   "source": [
    "from numpy import *\n",
    "\n",
    "def findLCS(v, w):\n",
    "    score = zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    backt = zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    for i in range(1,len(v)+1):\n",
    "        for j in range(1,len(w)+1):\n",
    "            # find best score at each vertex\n",
    "            if (v[i-1] == w[j-1]):\n",
    "                score[i,j], backt[i,j] = max((score[i-1,j-1]+1,3), (score[i-1,j],1), (score[i,j-1],2))\n",
    "            else:\n",
    "                score[i,j], backt[i,j] = max((score[i-1,j],1), (score[i,j-1],2))\n",
    "    return score, backt\n",
    "s, b = findLCS(v,w)\n",
    "print(\"s\\n\",s)\n",
    "print(\"b\\n\",b)\n",
    "# for i in range(len(s)):\n",
    "#     print(\"%10s %-20s    %12s %-20s\" % ('' if i else 'score =', str(s[i]), '' if i else 'backtrack =', str(b[i])))\n",
    "\n",
    "def LCS(b,v,i,j):\n",
    "    print(i,j)\n",
    "    if ((i == 0) or (j == 0)):\n",
    "        return ''\n",
    "    elif (b[i,j] == 3):\n",
    "        return LCS(b,v,i-1,j-1) + v[i-1]\n",
    "    elif (b[i,j] == 2):\n",
    "        return LCS(b,v,i,j-1)\n",
    "    else:\n",
    "        return LCS(b,v,i-1,j)\n",
    "# print(b.shape)\n",
    "# print(s.shape)\n",
    "print(LCS(b,v,b.shape[0]-1,b.shape[1]-1))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 2,
    "section": "answer"
   },
   "source": [
    "LCS = AACACAAACTGAGCCCCTACCCCAAATTAGACTGGATCTGGACCCTTCCATTCATTAGTCAGGGGGTCCTTCATTTTACCAATGT\n",
    "edit_distance = 85\n",
    "\n",
    "\n",
    "code:\n",
    "\n",
    "from numpy import *\n",
    "\n",
    "def findLCS(v, w):\n",
    "    score = zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    backt = zeros((len(v)+1,len(w)+1), dtype=\"int32\")\n",
    "    for i in range(1,len(v)+1):\n",
    "        for j in range(1,len(w)+1):\n",
    "            # find best score at each vertex\n",
    "            if (v[i-1] == w[j-1]):\n",
    "                score[i,j], backt[i,j] = max((score[i-1,j-1]+1,3), (score[i-1,j],1), (score[i,j-1],2))\n",
    "            else:\n",
    "                score[i,j], backt[i,j] = max((score[i-1,j],1), (score[i,j-1],2))\n",
    "    return score, backt\n",
    "s, b = findLCS(v,w)\n",
    "print(\"s\\n\",s)\n",
    "print(\"b\\n\",b)\n",
    "# for i in range(len(s)):\n",
    "#     print(\"%10s %-20s    %12s %-20s\" % ('' if i else 'score =', str(s[i]), '' if i else 'backtrack =', str(b[i])))\n",
    "\n",
    "def LCS(b,v,i,j):\n",
    "    print(i,j)\n",
    "    if ((i == 0) or (j == 0)):\n",
    "        return ''\n",
    "    elif (b[i,j] == 3):\n",
    "        return LCS(b,v,i-1,j-1) + v[i-1]\n",
    "    elif (b[i,j] == 2):\n",
    "        return LCS(b,v,i,j-1)\n",
    "    else:\n",
    "        return LCS(b,v,i-1,j)\n",
    "# print(b.shape)\n",
    "# print(s.shape)\n",
    "print(LCS(b,v,b.shape[0]-1,b.shape[1]-1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 3,
    "section": "problem"
   },
   "source": [
    "\n",
    "---\n",
    "**Problem #3.** (10 points) Suppose that you are given the following *constrained global alignment problem* for the sequences $v$ of length $n$, and $w$ of length $m$, which forces $v_i$ to be aligned with $w_j$. \n",
    "\n",
    "$$\\begin{aligned}\n",
    "v_1, v_2, ... v_{i-1}, &v_i, v_{i+1}, ... v_n \\\\\n",
    "                       &\\:|  \\\\\n",
    "w_1, w_2, ... w_{j-1}, &w_j, w_{j+1}, ... w_m \\\\\n",
    "\\end{aligned}$$\n",
    "\n",
    "How does the choice of $i$ and $j$ impact the performace of the algorithm? What choice of $i, j$ leads to the greatest speed up over a standard global alignment?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 3,
    "section": "answer"
   },
   "source": [
    "We can treat the contrained global alignment problem as a potential divide and conquer algorithm problem. \n",
    "\n",
    "This allows us to choose values of i and j that allow us to effect the performance of the algorithm. The closer that (i,j) is to where the best scoring path crosses the midpoint of the table, the better our algorithm's performance will be. Thus, selecting (i,j) at the point where the optimal path crosses the middle column of the table will allow us to have the greatest speedup over a standard global alignment. Additionally, selecting i and j as simply the midpoints of each of the sequences may help. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 4,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #4:** (5 points) Consider the following three peptide sequences:\n",
    "\n",
    "        p1 = \"KDPLTNEWKGPDPVLIW\"\n",
    "        p2 = \"WILVPDPGKDPLTNEWK\"\n",
    "        p3 = \"NEWKGPDPVLIWKDPLT\"\n",
    "\n",
    "What fraction of *each* petide's mass peaks are shared by all three peptides?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P1 Spec\n",
      " [57, 97, 99, 101, 113, 114, 115, 128, 129, 154, 185, 186, 196, 210, 212, 214, 215, 226, 243, 269, 282, 299, 309, 311, 314, 315, 325, 328, 340, 344, 366, 371, 397, 408, 412, 422, 424, 425, 426, 429, 443, 453, 457, 465, 468, 494, 500, 511, 521, 530, 537, 540, 554, 557, 578, 583, 593, 597, 608, 614, 634, 643, 658, 668, 669, 680, 691, 706, 711, 712, 715, 723, 740, 771, 779, 797, 809, 812, 819, 820, 826, 828, 855, 868, 877, 892, 908, 923, 925, 927, 983, 1005, 1021, 1022, 1024, 1040, 1111, 1123, 1134, 1135, 1137, 1168, 1191, 1234, 1236, 1248, 1252, 1265, 1320, 1333, 1349, 1380, 1434, 1446, 1448, 1462, 1477, 1535, 1559, 1561, 1576, 1648, 1674, 1689, 1745, 1802, 1860, 1988] 128\n",
      "P2 Spec\n",
      " [57, 97, 99, 101, 113, 114, 115, 128, 129, 154, 185, 186, 196, 210, 212, 214, 215, 226, 243, 269, 282, 299, 300, 309, 311, 314, 315, 325, 328, 340, 344, 366, 397, 408, 412, 422, 424, 425, 426, 429, 443, 453, 457, 465, 494, 510, 511, 512, 521, 530, 537, 540, 554, 557, 578, 593, 607, 608, 609, 611, 634, 643, 658, 668, 669, 691, 706, 708, 722, 723, 725, 740, 771, 797, 805, 819, 820, 821, 822, 823, 854, 855, 868, 877, 918, 920, 934, 937, 951, 983, 1005, 1019, 1031, 1034, 1040, 1066, 1111, 1120, 1132, 1133, 1137, 1144, 1163, 1168, 1217, 1245, 1246, 1252, 1262, 1265, 1330, 1349, 1359, 1375, 1380, 1431, 1448, 1477, 1488, 1545, 1561, 1576, 1674, 1689, 1802, 1860, 1988] 127\n",
      "P3 Spec\n",
      " [57, 97, 99, 101, 113, 114, 115, 128, 129, 154, 185, 186, 196, 210, 212, 214, 226, 243, 269, 282, 299, 309, 311, 314, 315, 325, 340, 366, 371, 397, 408, 412, 422, 424, 426, 427, 429, 443, 453, 465, 468, 494, 500, 511, 521, 526, 537, 540, 542, 554, 557, 578, 583, 593, 597, 608, 614, 634, 639, 655, 680, 691, 706, 711, 712, 723, 736, 740, 752, 754, 779, 809, 819, 820, 826, 851, 853, 865, 877, 892, 908, 923, 948, 964, 966, 1005, 1021, 1022, 1061, 1063, 1065, 1120, 1133, 1134, 1135, 1160, 1162, 1176, 1191, 1217, 1248, 1273, 1277, 1319, 1320, 1330, 1345, 1374, 1431, 1434, 1448, 1458, 1531, 1559, 1562, 1563, 1644, 1660, 1677, 1745, 1773, 1774, 1874, 1887, 1988] 125\n",
      "Frac Shared\n",
      " {128, 129, 521, 269, 397, 408, 537, 282, 154, 412, 540, 422, 424, 1448, 426, 299, 554, 429, 557, 691, 819, 309, 820, 311, 185, 186, 57, 314, 315, 443, 578, 706, 196, 325, 453, 1988, 465, 210, 593, 212, 340, 214, 723, 608, 97, 226, 99, 740, 101, 877, 366, 494, 1005, 113, 114, 115, 243, 634, 511} 59\n",
      "Shared P1: 0.4609375\n",
      "Shared P2: 0.4645669291338583\n",
      "Shared P3: 0.472\n"
     ]
    }
   ],
   "source": [
    "p1 = \"KDPLTNEWKGPDPVLIW\"\n",
    "p2 = \"WILVPDPGKDPLTNEWK\"\n",
    "p3 = \"NEWKGPDPVLIWKDPLT\"\n",
    "Daltons = { \n",
    "    'A':  71, 'C': 103, 'D': 115, 'E': 129,\n",
    "    'F': 147, 'G':  57, 'H': 137, 'I': 113,\n",
    "    'K': 128, 'L': 113, 'M': 131, 'N': 114,\n",
    "    'P':  97, 'Q': 128, 'R': 156, 'S':  87,\n",
    "    'T': 101, 'V':  99, 'W': 186, 'Y': 163 \n",
    "}\n",
    "\n",
    "\n",
    "def TheoreticalSpectrum(peptide):\n",
    "    # Generate every possible fragment of a peptide\n",
    "    spectrum = set()\n",
    "    for fragLength in range(1,len(peptide)+1):\n",
    "        for start in range(0,len(peptide)-fragLength+1):\n",
    "            seq = peptide[start:start+fragLength]\n",
    "            spectrum.add(sum([Daltons[res] for res in seq]))\n",
    "    return sorted(spectrum)\n",
    "\n",
    "p1_spec = TheoreticalSpectrum(p1)\n",
    "print(\"P1 Spec\\n\",p1_spec, len(p1_spec))\n",
    "\n",
    "p2_spec = TheoreticalSpectrum(p2)\n",
    "print(\"P2 Spec\\n\",p2_spec, len(p2_spec))\n",
    "\n",
    "p3_spec = TheoreticalSpectrum(p3)\n",
    "print(\"P3 Spec\\n\",p3_spec, len(p3_spec))\n",
    "\n",
    "\n",
    "P1_P2 = set(p1_spec).intersection(set(p2_spec))\n",
    "# print(\"set\",len(set(p1_spec)))\n",
    "# print(P1_P2, len(P1_P2))\n",
    "\n",
    "frac_shared = set(p1_spec).intersection(set(p2_spec)).intersection(p3_spec)\n",
    "print(\"Frac Shared\\n\",frac_shared, len(frac_shared))\n",
    "print(\"Shared P1:\", len(frac_shared)/len(p1_spec))\n",
    "print(\"Shared P2:\", len(frac_shared)/len(p2_spec))\n",
    "print(\"Shared P3:\", len(frac_shared)/len(p3_spec))\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 4,
    "section": "answer"
   },
   "source": [
    "Shared P1: 0.4609375\n",
    "Shared P2: 0.4645669291338583\n",
    "Shared P3: 0.472\n",
    "\n",
    "Code:\n",
    "\n",
    "p1 = \"KDPLTNEWKGPDPVLIW\"\n",
    "p2 = \"WILVPDPGKDPLTNEWK\"\n",
    "p3 = \"NEWKGPDPVLIWKDPLT\"\n",
    "\n",
    "def TheoreticalSpectrum(peptide):\n",
    "    # Generate every possible fragment of a peptide\n",
    "    spectrum = set()\n",
    "    for fragLength in range(1,len(peptide)+1):\n",
    "        for start in range(0,len(peptide)-fragLength+1):\n",
    "            seq = peptide[start:start+fragLength]\n",
    "            spectrum.add(sum([Daltons[res] for res in seq]))\n",
    "    return sorted(spectrum)\n",
    "\n",
    "p1_spec = TheoreticalSpectrum(p1)\n",
    "print(\"P1 Spec\\n\",p1_spec, len(p1_spec))\n",
    "\n",
    "p2_spec = TheoreticalSpectrum(p2)\n",
    "print(\"P2 Spec\\n\",p2_spec, len(p2_spec))\n",
    "\n",
    "p3_spec = TheoreticalSpectrum(p3)\n",
    "print(\"P3 Spec\\n\",p3_spec, len(p3_spec))\n",
    "\n",
    "\n",
    "P1_P2 = set(p1_spec).intersection(set(p2_spec))\n",
    "# print(\"set\",len(set(p1_spec)))\n",
    "# print(P1_P2, len(P1_P2))\n",
    "\n",
    "frac_shared = set(p1_spec).intersection(set(p2_spec)).intersection(p3_spec)\n",
    "print(\"Frac Shared\\n\",frac_shared, len(frac_shared))\n",
    "print(\"Shared P1:\", len(frac_shared)/len(p1_spec))\n",
    "print(\"Shared P2:\", len(frac_shared)/len(p2_spec))\n",
    "print(\"Shared P3:\", len(frac_shared)/len(p3_spec))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 5,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #5:** (5 points) Of the three peptide sequences given in *Problem #2*, which has the fewest private peaks, and how many does it have?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1024, 512, 510, 1031, 521, 1034, 1545, 526, 1040, 530, 1559, 1531, 537, 1561, 1562, 540, 1563, 542, 1061, 1063, 1576, 1065, 554, 1535, 1066, 557, 57, 578, 583, 593, 597, 1111, 607, 608, 97, 609, 99, 1123, 101, 614, 611, 1120, 1132, 1133, 1134, 1135, 1648, 113, 114, 115, 1137, 1644, 1144, 634, 1660, 639, 128, 129, 643, 1160, 1674, 1163, 1162, 1677, 655, 1168, 658, 1176, 1689, 154, 668, 669, 1191, 680, 691, 185, 186, 1217, 706, 196, 708, 711, 712, 715, 1745, 210, 723, 212, 1234, 214, 215, 1236, 722, 725, 1245, 1246, 1248, 736, 226, 740, 1252, 1773, 1262, 1774, 752, 1265, 754, 243, 1273, 1277, 771, 1802, 779, 269, 282, 797, 805, 1319, 1320, 809, 299, 812, 300, 1330, 819, 820, 309, 1333, 311, 821, 822, 314, 315, 826, 828, 823, 1345, 1860, 325, 1349, 328, 1359, 1874, 851, 340, 853, 854, 855, 344, 1374, 1375, 1887, 865, 868, 1380, 877, 366, 371, 892, 908, 397, 918, 1431, 408, 920, 1434, 923, 412, 925, 927, 422, 1446, 424, 425, 426, 1448, 934, 429, 937, 427, 1458, 948, 1462, 951, 443, 1988, 453, 1477, 964, 966, 457, 1488, 465, 468, 983, 1005, 494, 500, 1019, 1021, 1022, 511} 202\n",
      "Diff P1:\n",
      " {512, 1031, 1160, 1545, 1034, 1163, 1162, 1677, 526, 655, 1019, 918, 1431, 1176, 920, 1562, 1563, 542, 1061, 805, 1063, 1319, 1065, 1066, 934, 300, 937, 427, 1330, 1458, 948, 821, 822, 823, 951, 1217, 1345, 708, 964, 966, 1359, 1488, 722, 1874, 851, 725, 853, 854, 1245, 1246, 607, 1120, 609, 736, 611, 1374, 1375, 1887, 865, 1132, 1133, 1644, 1773, 1262, 1774, 752, 754, 1144, 1273, 1531, 1660, 1277, 510, 639} 74\n",
      "Diff P2:\n",
      " {1024, 1160, 892, 1162, 779, 908, 1677, 526, 655, 1021, 1559, 1176, 1562, 1563, 1434, 923, 542, 925, 927, 1061, 1446, 1063, 1191, 1065, 680, 1535, 1319, 1320, 809, 812, 427, 1458, 948, 1333, 1462, 826, 828, 1345, 964, 966, 583, 711, 712, 715, 1745, 1234, 1874, 1236, 597, 851, 853, 468, 1374, 1887, 1248, 736, 865, 1123, 614, 1644, 1773, 1134, 1135, 1648, 1774, 752, 754, 371, 500, 1273, 1531, 1660, 1277, 1022, 639} 75\n",
      "Diff P3:\n",
      " {1024, 512, 643, 771, 1031, 1545, 1034, 1674, 1163, 1802, 1040, 1168, 530, 658, 918, 920, 1561, 1689, 668, 669, 797, 925, 927, 805, 1446, 934, 1576, 425, 1066, 937, 812, 300, 1333, 821, 822, 823, 1462, 951, 828, 708, 1860, 1349, 1477, 328, 457, 715, 1359, 1488, 1234, 722, 1236, 725, 854, 1111, 215, 855, 344, 983, 1245, 1246, 607, 1375, 609, 1123, 611, 1252, 868, 1380, 1132, 1262, 1648, 1137, 1265, 1144, 1019, 510, 1535} 77\n"
     ]
    }
   ],
   "source": [
    "union_peaks = set(p1_spec)|set(p2_spec)|set(p3_spec)\n",
    "print(union_peaks, len(union_peaks))\n",
    "\n",
    "\n",
    "p1_diff = union_peaks-set(p1_spec)\n",
    "print(\"Diff P1:\\n\", p1_diff, len(p1_diff))\n",
    "p2_diff = union_peaks-set(p2_spec)\n",
    "print(\"Diff P2:\\n\", p2_diff, len(p2_diff))\n",
    "p3_diff = union_peaks-set(p3_spec)\n",
    "print(\"Diff P3:\\n\", p3_diff, len(p3_diff))\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 5,
    "section": "answer"
   },
   "source": [
    "private --> only one entity posseses\n",
    "I used set difference. I found the union of all the mass spectrums, then found the difference of each individual spectrum with the union of all the spectrums. \n",
    "Diff P1: 74\n",
    "Diff P2: 75\n",
    "Diff P3: 77\n",
    "\n",
    "This means p1 has 74 private peaks, the least of the three peptide sequences. \n",
    "\n",
    "Code:\n",
    "\n",
    "union_peaks = set(p1_spec)|set(p2_spec)|set(p3_spec)\n",
    "print(union_peaks, len(union_peaks))\n",
    "\n",
    "\n",
    "p1_diff = union_peaks-set(p1_spec)\n",
    "print(\"Diff P1:\\n\", p1_diff, len(p1_diff))\n",
    "p2_diff = union_peaks-set(p2_spec)\n",
    "print(\"Diff P2:\\n\", p2_diff, len(p2_diff))\n",
    "p3_diff = union_peaks-set(p3_spec)\n",
    "print(\"Diff P3:\\n\", p3_diff, len(p3_diff))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 6,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem 6:** (5 points) Given the following experimental spectrum:\n",
    "\n",
    "    experimental = [57, 97, 99, 101, 113, 114, 115, 128, 129, 154, 185, 186, 196, \n",
    "                    210, 212, 214, 215, 226, 243, 269, 299, 309, 311, 314, 315, 325, \n",
    "                    328, 340, 344, 366, 371, 397, 408, 412, 422, 424, 425, 426, 429,\n",
    "                    443, 453, 457, 465, 468, 494, 500, 511, 521, 530, 537, 540, 554, \n",
    "                    557, 578, 583, 593, 597, 608, 614, 634, 643, 658, 668, 669, 680,\n",
    "                    691, 706, 711, 712, 715, 723, 740, 771, 779, 797, 809, 812, 819,\n",
    "                    820, 826, 828, 868, 877, 892, 908, 923, 925, 927, 983, 1005, 1021, \n",
    "                    1022, 1024, 1030, 1040, 1111, 1123, 1134, 1135, 1168, 1191, 1234, \n",
    "                    1236, 1250, 1252, 1265, 1320, 1328, 1333, 1349, 1380, 1434, 1446, \n",
    "                    1448, 1462, 1477, 1535, 1559, 1576, 1648, 1674, 1689, 1743, 1745, \n",
    "                    1860, 1988]\n",
    "\n",
    "Which of the three given peptide sequences most likely produced it? Explain your reasoning."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Diff p1\n",
      " 4\n",
      "Diff p1\n",
      " 6\n",
      "Diff p2\n",
      " 42\n",
      "Diff p2\n",
      " 43\n",
      "Diff p3\n",
      " 44\n",
      "Diff p3\n",
      " 43\n"
     ]
    }
   ],
   "source": [
    "experimental = [57, 97, 99, 101, 113, 114, 115, 128, 129, 154, 185, 186, 196, \n",
    "                210, 212, 214, 215, 226, 243, 269, 299, 309, 311, 314, 315, 325, \n",
    "                328, 340, 344, 366, 371, 397, 408, 412, 422, 424, 425, 426, 429,\n",
    "                443, 453, 457, 465, 468, 494, 500, 511, 521, 530, 537, 540, 554, \n",
    "                557, 578, 583, 593, 597, 608, 614, 634, 643, 658, 668, 669, 680,\n",
    "                691, 706, 711, 712, 715, 723, 740, 771, 779, 797, 809, 812, 819,\n",
    "                820, 826, 828, 868, 877, 892, 908, 923, 925, 927, 983, 1005, 1021, \n",
    "                1022, 1024, 1030, 1040, 1111, 1123, 1134, 1135, 1168, 1191, 1234, \n",
    "                1236, 1250, 1252, 1265, 1320, 1328, 1333, 1349, 1380, 1434, 1446, \n",
    "                1448, 1462, 1477, 1535, 1559, 1576, 1648, 1674, 1689, 1743, 1745, \n",
    "                1860, 1988]\n",
    "\n",
    "exp = set(experimental)\n",
    "p1_s = set(p1_spec)\n",
    "p2_s = set(p2_spec)\n",
    "p3_s = set(p3_spec)\n",
    "\n",
    "print(\"Diff p1\\n\", len(exp-p1_s))\n",
    "print(\"Diff p1\\n\", len(p1_s-exp))\n",
    "\n",
    "print(\"Diff p2\\n\", len(exp-p2_s))\n",
    "print(\"Diff p2\\n\", len(p2_s-exp))\n",
    "\n",
    "print(\"Diff p3\\n\",len(exp-p3_s))\n",
    "print(\"Diff p3\\n\", len(p3_s-exp))\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 6,
    "section": "answer"
   },
   "source": [
    "p1\n",
    "\n",
    "It is most likely that p1 produced it, since there are very few differences between its spectrum and the experimental spectrum. \n",
    "\n",
    "\n",
    "Diff experimental-p1, p1-experimental\n",
    " 4,6\n",
    "Diff experimental-p2, p2-experimental\n",
    " 42,43\n",
    "Diff experimental-p3, p3-experimental\n",
    " 44,43\n",
    "\n",
    "Code:\n",
    "\n",
    "experimental = [57, 97, 99, 101, 113, 114, 115, 128, 129, 154, 185, 186, 196, \n",
    "                210, 212, 214, 215, 226, 243, 269, 299, 309, 311, 314, 315, 325, \n",
    "                328, 340, 344, 366, 371, 397, 408, 412, 422, 424, 425, 426, 429,\n",
    "                443, 453, 457, 465, 468, 494, 500, 511, 521, 530, 537, 540, 554, \n",
    "                557, 578, 583, 593, 597, 608, 614, 634, 643, 658, 668, 669, 680,\n",
    "                691, 706, 711, 712, 715, 723, 740, 771, 779, 797, 809, 812, 819,\n",
    "                820, 826, 828, 868, 877, 892, 908, 923, 925, 927, 983, 1005, 1021, \n",
    "                1022, 1024, 1030, 1040, 1111, 1123, 1134, 1135, 1168, 1191, 1234, \n",
    "                1236, 1250, 1252, 1265, 1320, 1328, 1333, 1349, 1380, 1434, 1446, \n",
    "                1448, 1462, 1477, 1535, 1559, 1576, 1648, 1674, 1689, 1743, 1745, \n",
    "                1860, 1988]\n",
    "\n",
    "exp = set(experimental)\n",
    "p1_s = set(p1_spec)\n",
    "p2_s = set(p2_spec)\n",
    "p3_s = set(p3_spec)\n",
    "\n",
    "print(\"Diff p1\\n\",exp-p1_s, len(exp-p1_s))\n",
    "print(\"Diff p1\\n\",p1_s-exp, len(p1_s-exp))\n",
    "\n",
    "print(\"Diff p2\\n\",exp-p2_s, len(exp-p2_s))\n",
    "print(\"Diff p2\\n\",p2_s-exp, len(p2_s-exp))\n",
    "\n",
    "print(\"Diff p3\\n\",exp-p3_s, len(exp-p3_s))\n",
    "print(\"Diff p3\\n\",p3_s-exp, len(p3_s-exp))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 7,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #7.** (5 points) The following question assumes the definitions of *breakpoint* and *strip* used in the discussion of reversal sorting from Lectures 24 to 25. Moreover, you should assume all permutations are extended (i.e. sandwiched between the fixed elements 0 and n+1). Is it possible to have a permutation of ***N*** elements with exactly 2 breakpoints where those breakpoints are both sides of a single element? Give an example, or explain why not."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 7,
    "section": "answer"
   },
   "source": [
    "\n",
    "In lectures 24 to 25, we define breakpoints and strips for a permutation of N generic orderable elements. \n",
    "\n",
    "If we were to try to satisfy the base case (N=1), it would be impossible to have 2 breakpoints. By looking at larger cases (N>1) we find that any single element surrounded by two breakpoints will force there to be breakpoints next to the extensions. We can prove this by realizing that if there are exactly 2 breakpoints, there must be 3 strips, with one of the strips length 1. Since the sequence is not sorted yet, one of these strips must be reversed. This will force a breakpoint next to the fixed element making it impossible to have exactly 2 breakpoints where those breakpoints are both sides of a single element."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 8,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #8:** (10 points) Consider two genomes with homologous genes, represented as letters, in the following orders:\n",
    "```python\n",
    "    G1 = \"B,L,O,H,G,N,D,K,F,J,A,C,P,Q,M,E,I\"\n",
    "    G2 = \"C,B,A,D,P,K,N,L,E,H,G,M,O,F,I,Q,J\"\n",
    "```\n",
    "Find a series of reversals that transforms G1 to G2, where a reversal is given by the following function:\n",
    "```python\n",
    "    def reverse(genome,i,j):\n",
    "        gene = genome.split(',')\n",
    "        return ','.join(gene[:i] + [g for g in reversed(gene[i:j])] + gene[j:])\n",
    "```        \n",
    "Where the series of reversals\n",
    "```python\n",
    "    R1 = reverse(G1,1,11)\n",
    "    R2 = reverse(R1,0,11)\n",
    "    R3 = reverse(R2,0,12)\n",
    "    print(R3)\n",
    "```\n",
    "produces:\n",
    "\n",
    "    \"C,B,A,J,F,K,D,N,G,H,O,L,P,Q,M,E,I\"\n",
    "\n",
    "Provide your series of reversals below. More credit will be given for fewer reversals. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n",
      "{'C': 1, 'B': 2, 'A': 3, 'D': 4, 'P': 5, 'K': 6, 'N': 7, 'L': 8, 'E': 9, 'H': 10, 'G': 11, 'M': 12, 'O': 13, 'F': 14, 'I': 15, 'Q': 16, 'J': 17}\n",
      "[2, 8, 13, 10, 11, 7, 4, 6, 14, 17, 3, 1, 5, 16, 12, 9, 15]\n",
      "\n",
      "Strips: [(4, 6)] [(1, 2), (2, 3), (3, 4), (6, 7), (7, 8), (8, 9), (9, 10), (10, 11), (11, 12), (12, 13), (13, 14), (14, 15), (15, 16), (16, 17), (17, 18)]\n",
      "2: [0, 2, 8, 13, 10, 11, 7, 4, 6, 14, 17, 3, 1, 5, 16, 12, 9, 15, 18]  rho(4, 16)\n",
      "Press Enter:\n",
      "Strips: [] [(1, 2), (2, 3), (3, 5), (5, 6), (6, 7), (7, 8), (8, 9), (9, 10), (10, 11), (11, 12), (12, 13), (13, 14), (14, 17), (17, 18)]\n",
      "2: [0, 2, 8, 13, 12, 16, 5, 1, 3, 17, 14, 6, 4, 7, 11, 10, 9, 15, 18]  rho(1, 8)\n",
      "Press Enter:\n",
      "Strips: [(4, 6), (7, 9)] [(2, 3), (3, 4), (6, 7), (9, 10), (10, 11), (11, 12), (12, 13), (13, 14), (14, 17), (17, 18)]\n",
      "1: [0, 1, 5, 16, 12, 13, 8, 2, 3, 17, 14, 6, 4, 7, 11, 10, 9, 15, 18]  rho(7, 17)\n",
      "Press Enter:\n",
      "Strips: [(4, 6), (6, 10)] [(2, 3), (3, 4), (10, 11), (11, 12), (12, 13), (13, 14), (14, 15), (15, 17), (17, 18)]\n",
      "2: [0, 1, 5, 16, 12, 13, 8, 9, 10, 11, 7, 4, 6, 14, 17, 3, 2, 15, 18]  rho(14, 18)\n",
      "Press Enter:\n",
      "Strips: [(4, 6), (6, 10), (13, 15), (15, 17)] [(2, 3), (3, 4), (10, 11), (11, 12), (12, 13)]\n",
      "1: [0, 1, 5, 16, 12, 13, 8, 9, 10, 11, 7, 4, 6, 14, 15, 2, 3, 17, 18]  rho(3, 13)\n",
      "Press Enter:\n",
      "Strips: [(2, 4), (13, 15), (15, 17)] [(4, 5), (5, 6), (6, 10), (10, 12), (12, 13)]\n",
      "1: [0, 1, 5, 6, 4, 7, 11, 10, 9, 8, 13, 12, 16, 14, 15, 2, 3, 17, 18]  rho(6, 10)\n",
      "Press Enter:\n",
      "Strips: [(2, 4), (5, 10), (13, 15), (15, 17)] [(4, 5), (10, 12), (12, 13)]\n",
      "1: [0, 1, 5, 6, 4, 7, 8, 9, 10, 11, 13, 12, 16, 14, 15, 2, 3, 17, 18]  rho(10, 12)\n",
      "Press Enter:\n",
      "Strips: [(2, 4), (5, 12), (13, 15), (15, 17)] [(4, 5), (12, 13)]\n",
      "1: [0, 1, 5, 6, 4, 7, 8, 9, 10, 11, 12, 13, 16, 14, 15, 2, 3, 17, 18]  rho(13, 15)\n",
      "Press Enter:\n",
      "Strips: [(2, 4), (5, 12), (15, 17)] [(4, 5), (12, 15)]\n",
      "1: [0, 1, 5, 6, 4, 7, 8, 9, 10, 11, 12, 13, 16, 15, 14, 2, 3, 17, 18]  rho(12, 15)\n",
      "Press Enter:\n",
      "Strips: [(2, 4), (5, 15), (15, 17)] [(4, 5)]\n",
      "1: [0, 1, 5, 6, 4, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 2, 3, 17, 18]  rho(5, 17)\n",
      "Press Enter:\n",
      "Strips: [(2, 4)] [(4, 7), (7, 17)]\n",
      "1: [0, 1, 5, 6, 4, 3, 2, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 17, 18]  rho(2, 7)\n",
      "Press Enter:\n",
      "Strips: [] [(5, 7), (7, 17)]\n",
      "1: [0, 1, 2, 3, 4, 6, 5, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 17, 18]  rho(5, 7)\n",
      "Press Enter:\n",
      "Strips: [] [(7, 17)]\n",
      "2: [0, 1, 2, 3, 4, 5, 6, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 17, 18]  rho(7, 17)\n",
      "Press Enter:\n",
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18] Sorted\n",
      "13\n"
     ]
    }
   ],
   "source": [
    "G1 = \"BLOHGNDKFJACPQMEI\"\n",
    "G2 = \"CBADPKNLEHGMOFIQJ\"\n",
    "print(len(G2))\n",
    "conv_dict = {}\n",
    "for i in range(len(G2)):\n",
    "    conv_dict[G2[i]] = i+1\n",
    "print(conv_dict)\n",
    "number_list = []\n",
    "for i in range(len(G1)):\n",
    "    number_list.append(conv_dict[G1[i]])\n",
    "print(number_list)\n",
    "def hasBreakpoints(seq):\n",
    "    \"\"\" returns True if sequences is not strictly increasing by 1 \"\"\"\n",
    "    for i in range(1, len(seq)):\n",
    "        if (seq[i] != seq[i-1] + 1):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def getStrips(seq):\n",
    "    \"\"\" find contained intervals where sequence is ordered, and return intervals\n",
    "    in as lists, increasing and decreasing. Single elements are considered\n",
    "    decreasing. \"Contained\" excludes the first and last interval. \"\"\"\n",
    "    deltas = [seq[i+1] - seq[i] for i in range(len(seq)-1)]\n",
    "    increasing = list()\n",
    "    decreasing = list()\n",
    "    start = 0\n",
    "    for i, diff in enumerate(deltas):\n",
    "        if (abs(diff) == 1) and (diff == deltas[start]):\n",
    "            continue\n",
    "        if (start > 0):\n",
    "            if deltas[start] == 1:\n",
    "                increasing.append((start, i+1))\n",
    "            else:\n",
    "                decreasing.append((start, i+1))\n",
    "        start = i+1\n",
    "    return increasing, decreasing\n",
    "def pickReversal(seq, strips):\n",
    "    \"\"\" test each decreasing interval to see if it leads to a reversal that\n",
    "    removes two breakpoints, otherwise, return a reversal that removes only one \"\"\"\n",
    "    for i, j in strips:\n",
    "        k = seq.index(seq[j-1]-1)\n",
    "        if (seq[k+1] + 1 == seq[j]):\n",
    "            # removes 2 breakpoints\n",
    "            return 2, (min(k+1, j), max(k+1, j))\n",
    "    # In the worst case we remove only one, but avoid the length \"1\" strips\n",
    "    for i, j in strips:\n",
    "        k = seq.index(seq[j-1]-1)\n",
    "        if (j - i > 1):\n",
    "            break\n",
    "    return 1, (min(k+1, j), max(k+1, j))\n",
    "\n",
    "def doReversal(seq,reversal):\n",
    "    i, j = reversal\n",
    "    return seq[:i] + [element for element in reversed(seq[i:j])] + seq[j:]\n",
    "\n",
    "def improvedBreakpointReversalSort(seq, verbose=True):\n",
    "    seq = [0] + seq + [max(seq)+1]                             # Extend sequence\n",
    "    N = 0\n",
    "    while hasBreakpoints(seq):\n",
    "        increasing, decreasing = getStrips(seq)\n",
    "        if len(decreasing) > 0:                                # pick a reversal that removes a decreasing strip\n",
    "            removed, reversal = pickReversal(seq, decreasing)\n",
    "        else:\n",
    "            removed, reversal = 0, increasing[0]               # No breakpoints can be removed\n",
    "        if verbose:\n",
    "            print(\"Strips:\", increasing, decreasing)\n",
    "            print(\"%d: %s  rho%s\" % (removed, seq, reversal))\n",
    "            input(\"Press Enter:\")\n",
    "        seq = doReversal(seq,reversal)\n",
    "        N += 1\n",
    "    if verbose:\n",
    "        print(seq, \"Sorted\")\n",
    "    return N\n",
    "\n",
    "# Also try: [1,9,3,4,7,8,2,6,5]\n",
    "print()\n",
    "print(improvedBreakpointReversalSort(number_list, verbose=True))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 8,
    "section": "answer"
   },
   "source": [
    "1. rho(4, 16) --> R1 = reverse(G1,4,16)\n",
    "2. rho(1, 8) --> R2 = reverse(R1,1,18)\n",
    "3. rho(7, 17) --> R3 = reverse(R2,7,17)\n",
    "4. rho(14, 18) --> R4 = reverse(R3,14,18)\n",
    "5. rho(3, 13) --> R5 = reverse(R4,3,13)\n",
    "6. rho(6, 10) --> R6 = reverse(R5,6,10)\n",
    "7. rho(10, 12) --> R7 = reverse(R6,10,12)\n",
    "8. rho(13, 15) --> R8 = reverse(R7,13,15)\n",
    "9. rho(12, 15) --> R9 = reverse(R8,12,15)\n",
    "10. rho(5, 17) --> R10 = reverse(R9,5,17)\n",
    "11. rho(2, 7) --> R11 = reverse(R10,2,7)\n",
    "12. rho(5, 7) --> R12 = reverse(R11,5,7)\n",
    "13. rho(7, 17) --> R13 = reverse(R12,7,17)\n",
    "\n",
    "_____________________________________________\n",
    "Code:\n",
    "\n",
    "G1 = \"BLOHGNDKFJACPQMEI\"\n",
    "G2 = \"CBADPKNLEHGMOFIQJ\"\n",
    "print(len(G2))\n",
    "conv_dict = {}\n",
    "for i in range(len(G2)):\n",
    "    conv_dict[G2[i]] = i+1\n",
    "print(conv_dict)\n",
    "number_list = []\n",
    "for i in range(len(G1)):\n",
    "    number_list.append(conv_dict[G1[i]])\n",
    "print(number_list)\n",
    "\n",
    "\n",
    "\n",
    "def hasBreakpoints(seq):\n",
    "    \"\"\" returns True if sequences is not strictly increasing by 1 \"\"\"\n",
    "    for i in range(1, len(seq)):\n",
    "        if (seq[i] != seq[i-1] + 1):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def getStrips(seq):\n",
    "    \"\"\" find contained intervals where sequence is ordered, and return intervals\n",
    "    in as lists, increasing and decreasing. Single elements are considered\n",
    "    decreasing. \"Contained\" excludes the first and last interval. \"\"\"\n",
    "    deltas = [seq[i+1] - seq[i] for i in range(len(seq)-1)]\n",
    "    increasing = list()\n",
    "    decreasing = list()\n",
    "    start = 0\n",
    "    for i, diff in enumerate(deltas):\n",
    "        if (abs(diff) == 1) and (diff == deltas[start]):\n",
    "            continue\n",
    "        if (start > 0):\n",
    "            if deltas[start] == 1:\n",
    "                increasing.append((start, i+1))\n",
    "            else:\n",
    "                decreasing.append((start, i+1))\n",
    "        start = i+1\n",
    "    return increasing, decreasing\n",
    "def pickReversal(seq, strips):\n",
    "    \"\"\" test each decreasing interval to see if it leads to a reversal that\n",
    "    removes two breakpoints, otherwise, return a reversal that removes only one \"\"\"\n",
    "    for i, j in strips:\n",
    "        k = seq.index(seq[j-1]-1)\n",
    "        if (seq[k+1] + 1 == seq[j]):\n",
    "            # removes 2 breakpoints\n",
    "            return 2, (min(k+1, j), max(k+1, j))\n",
    "    # In the worst case we remove only one, but avoid the length \"1\" strips\n",
    "    for i, j in strips:\n",
    "        k = seq.index(seq[j-1]-1)\n",
    "        if (j - i > 1):\n",
    "            break\n",
    "    return 1, (min(k+1, j), max(k+1, j))\n",
    "\n",
    "def doReversal(seq,reversal):\n",
    "    i, j = reversal\n",
    "    return seq[:i] + [element for element in reversed(seq[i:j])] + seq[j:]\n",
    "\n",
    "def improvedBreakpointReversalSort(seq, verbose=True):\n",
    "    seq = [0] + seq + [max(seq)+1]                             # Extend sequence\n",
    "    N = 0\n",
    "    while hasBreakpoints(seq):\n",
    "        increasing, decreasing = getStrips(seq)\n",
    "        if len(decreasing) > 0:                                # pick a reversal that removes a decreasing strip\n",
    "            removed, reversal = pickReversal(seq, decreasing)\n",
    "        else:\n",
    "            removed, reversal = 0, increasing[0]               # No breakpoints can be removed\n",
    "        if verbose:\n",
    "            print(\"Strips:\", increasing, decreasing)\n",
    "            print(\"%d: %s  rho%s\" % (removed, seq, reversal))\n",
    "            input(\"Press Enter:\")\n",
    "        seq = doReversal(seq,reversal)\n",
    "        N += 1\n",
    "    if verbose:\n",
    "        print(seq, \"Sorted\")\n",
    "    return N\n",
    "\n",
    "# Also try: [1,9,3,4,7,8,2,6,5]\n",
    "print('\\n\\n')\n",
    "print(improvedBreakpointReversalSort(number_list, verbose=True))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 9,
    "section": "problem"
   },
   "source": [
    "---\n",
    "The start of most mammalian genes are associated with promoter regions that exhbit a higher than expected occurance of *CG* dinucleotides. These regions are called ***\"CG-islands\"***. Assume a simplified hidden-Markov model of a genome which alternates betten two states, normal (<b>N</b>) and CG-island (<b>I</b>), as depicted below:  \n",
    "\n",
    "<div align=\"center\"><img src=\"http://csbio.unc.edu/mcmillan/Comp555S18/Media/FinalCGHMM.png\" width=\"400px\"></div>\n",
    "\n",
    "---\n",
    "**Problem #9.** (10 points) How many Cs or Gs within a window of length *n* would imply that you are most likely within a CG-island?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 9,
    "section": "answer"
   },
   "source": [
    "\n",
    "We know that : (2/5)^k * (3/5)^(n-k)  =  (4/5)^k * (1/5)^(n-k)\n",
    "\n",
    "Using WolframAlpha this simplifies to: n = (k (log(2) + log(3)))/log(3), or n = 1.630929753571457437099527114342*k\n",
    "\n",
    "This implies k = 0.6131471927654584131297538615321791235348*n\n",
    "\n",
    "This makes sense since while the hidden state I is in place, there is a 4/5 chance of seeing a C or G.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 10,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem 10.** (5 points) The following graph shows the Log-Odds of $P(b_i | N)$ over $P(b_i | I)$ for the sequence shown along the x-axis computed in two ways. \n",
    "<br><br>\n",
    "\n",
    "<div align=\"center\"><img src=\"http://csbio.unc.edu/mcmillan/Comp555S18/Media/FinalLogOdds.png\" width=\"800px\"></div>\n",
    "\n",
    "The blue line is based on ratios from the Forward-Backward solution at every base, and the gold line is based on a 7-base window (extending 3-bases on both sides of the plotted point) assuming a state sequence of \"NNNNNNN\" and \"IIIIIII\". Use this inforamtion to predict the hidden state for each base of the given 50-base sequence.  "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 10,
    "section": "answer"
   },
   "source": [
    "\n",
    "TCATA|ATCTG|GGGGT|CGGCG|CGGGG|GGCCG|GACGG|CTCAC|GATGT|ACTTG\n",
    "\n",
    "Prediction:\n",
    "\n",
    "NNNNN NNNNN NNIII IIIII IIIII IIIII IIIII INNNN NNNNN NNNNN\n",
    "TCATA|ATCTG|GGGGT|CGGCG|CGGGG|GGCCG|GACGG|CTCAC|GATGT|ACTTG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 11,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem 11.** (5 points) How does the likelihood of your state predictions from problem 10 compare to the actual series of states used to generate the given base sequence shown below?\n",
    "\n",
    "                             |    |    |    |    |    |    |    |    |    |\n",
    "            sequence:    TCATAATCTGGGGGTCGGCGCGGGGGGCCGGACGGCTCACGATGTACTTG\n",
    "            NI-state:    NNNNNNNNNNNIIIIIIIIIIIIIIIIIIIIIIIIIINNNNNNNNNIIII\n",
    "\n",
    "What does this imply?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 11,
    "section": "answer"
   },
   "source": [
    "                   |    |    |    |    |    |    |    |    |    |\n",
    "Sequence:      TCATAATCTGGGGGTCGGCGCGGGGGGCCGGACGGCTCACGATGTACTTG\n",
    "NI-state:      NNNNNNNNNNNIIIIIIIIIIIIIIIIIIIIIIIIIINNNNNNNNNIIII\n",
    "Prediction:    NNNNNNNNNNNNIIIIIIIIIIIIIIIIIIIIIIIINNNNNNNNNNNNNN\n",
    "\n",
    "My prediction based off of the Logs-Odd Ratios fits pretty well. However, at the end the actual series of states switches to I, which was not easy to predict based off of the Logs-Odd Ratio plot. This makes some sense because the blue plot is trending downward towards 0, meaning that it is difficult to confidently predict the state. The closer the Ratio is to the middle (0.0) the less certain our prediction of the state can be. \n",
    "\n",
    "This may imply that even though the state was \"I\", it was \"acting\" like an \"N\", because there is an unlikely probability that it may not act as expected for brief periods of times. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 12,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #12.** (5 points) Produce a series of state predictions that is *unlikely* to have produced the given base sequence from problems 10 and 11. Explain why?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NINNIIINIIIINNNNNIIININNIIIINNNNNNIIIINNNIIIIINIINN\n"
     ]
    }
   ],
   "source": [
    "NI_list = ['N', 'I', 'N', 'N', 'I', 'I', 'I', 'N', 'I', 'I', 'I', 'I', 'N', 'N', 'N', 'N', 'N', 'I', 'I', 'I', 'N', 'I', 'N', 'N', 'I', 'I', 'I', 'I', 'N', 'N', 'N', 'N', 'N', 'N', 'I', 'I', 'I', 'I', 'N', 'N', 'N', 'I', 'I', 'I', 'I', 'I', 'N', 'I', 'I', 'N', 'N']\n",
    "NI_s = \"\"\n",
    "for i in NI_list:\n",
    "    NI_s+= i\n",
    "print(NI_s)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 12,
    "section": "answer"
   },
   "source": [
    "NI-state:      NINNIIINIIIINNNNNIIININNIIIINNNNNIIIINNNIIIIINIINN\n",
    "sequence:      TCATAATCTGGGGGTCGGCGCGGGGGGCCGGACGGCTCACGATGTACTTG\n",
    "\n",
    "\n",
    "Using Lecture 22, Slide 22, I brute forced the veterbi problem to produce an unlikely prediction.\n",
    "\n",
    "Code:\n",
    "['N', 'I', 'N', 'N', 'I', 'I', 'I', 'N', 'I', 'I', 'I', 'I', 'N', 'N', 'N', 'N', 'N', 'I', 'I', 'I', 'N', 'I', 'N', 'N', 'I', 'I', 'I', 'I', 'N', 'N', 'N', 'N', 'N', 'N', 'I', 'I', 'I', 'I', 'N', 'N', 'N', 'I', 'I', 'I', 'I', 'I', 'N', 'I', 'I', 'N', 'N']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 13,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #13.** (5 points) What is the smallest value of <i>k</i> where the number of possible <i>k</i>-mer motif positions within 10 DNA strands of length 100 is less than the number of possible <i>k</i>-mers? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmerCounts(seq, k):\n",
    "    kmerDict = {}\n",
    "    for i in range(1,len(seq)-k+1):\n",
    "        kmer = seq[i:i+k]\n",
    "        kmerDict[kmer] = kmerDict.get(kmer,0) + 1\n",
    "    return kmerDict\n",
    "\n",
    "print('  k     k-mers              4^k      N-k+1          missing   repeated')\n",
    "for k in range(2,25):\n",
    "    kmers = kmerCounts(seq, k)\n",
    "    print(\"%3d %10d %16d %10d %16d %10d\" % (k, len(kmers), 4**k, (len(seq)-1)-k+1, 4**k-len(kmers), (len(seq)-1)-k+1-len(kmers)))\n",
    "    \n",
    "kmers = kmerCounts(seq, 4)\n",
    "    \n",
    "def printAllKLength(set, k): \n",
    "    n = len(set)  \n",
    "    printAllKLengthRec(set, \"\", n, k) \n",
    "  \n",
    "def printAllKLengthRec(set, prefix, n, k): \n",
    "    # Base case: k is 0, \n",
    "    # print prefix \n",
    "    if (k == 0) : \n",
    "#         print(prefix)\n",
    "        all_kmers.append(prefix)\n",
    "        return\n",
    "  \n",
    "    # One by one add all characters  \n",
    "    # from set and recursively  \n",
    "    # call for k equals to k-1 \n",
    "    for i in range(n): \n",
    "  \n",
    "        # Next character of input added \n",
    "        newPrefix = prefix + set[i] \n",
    "          \n",
    "        # k is decreased, because  \n",
    "        # we have added a new character \n",
    "        printAllKLengthRec(set, newPrefix, n, k - 1)\n",
    "all_kmers = []\n",
    "set1 = ['A','C','G','T']\n",
    "printAllKLength(set1, 4)\n",
    "# print(all_kmers)\n",
    "print(\"\\n\\n4 missing kmers:\")\n",
    "print(list(set(all_kmers) - set(kmers)))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 13,
    "section": "answer"
   },
   "source": [
    "k = 5\n",
    "\n",
    "After k = 5,  the number of possible k-mer motif positions within 10 DNA strands of length 100 is greater than the number of possible k-mers.\n",
    "\n",
    "Code:\n",
    "\n",
    "def kmerCounts(seq, k):\n",
    "    kmerDict = {}\n",
    "    for i in range(1,len(seq)-k+1):\n",
    "        kmer = seq[i:i+k]\n",
    "        kmerDict[kmer] = kmerDict.get(kmer,0) + 1\n",
    "    return kmerDict\n",
    "\n",
    "print('  k     k-mers              4^k      N-k+1          missing   repeated')\n",
    "for k in range(2,25):\n",
    "    kmers = kmerCounts(seq, k)\n",
    "    print(\"%3d %10d %16d %10d %16d %10d\" % (k, len(kmers), 4**k, (len(seq)-1)-k+1, 4**k-len(kmers), (len(seq)-1)-k+1-len(kmers)))\n",
    "    \n",
    "kmers = kmerCounts(seq, 4)\n",
    "    \n",
    "def printAllKLength(set, k): \n",
    "    n = len(set)  \n",
    "    printAllKLengthRec(set, \"\", n, k) \n",
    "  \n",
    "def printAllKLengthRec(set, prefix, n, k): \n",
    "      \n",
    "    # Base case: k is 0, \n",
    "    # print prefix \n",
    "    if (k == 0) : \n",
    "#         print(prefix)\n",
    "        all_kmers.append(prefix)\n",
    "        return\n",
    "  \n",
    "    # One by one add all characters  \n",
    "    # from set and recursively  \n",
    "    # call for k equals to k-1 \n",
    "    for i in range(n): \n",
    "  \n",
    "        # Next character of input added \n",
    "        newPrefix = prefix + set[i] \n",
    "          \n",
    "        # k is decreased, because  \n",
    "        # we have added a new character \n",
    "        printAllKLengthRec(set, newPrefix, n, k - 1)\n",
    "all_kmers = []\n",
    "set1 = ['A','C','G','T']\n",
    "printAllKLength(set1, 4)\n",
    "# print(all_kmers)\n",
    "print(\"\\n\\n4 missing kmers:\")\n",
    "print(list(set(all_kmers) - set(kmers)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 14,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #14.** (5 points) Which of the following is larger an uncompressed msBWT that indexes 100 DNA sequences each of length 100 or the suffix array over the same 100 DNA sequences?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 14,
    "section": "answer"
   },
   "source": [
    "The suffix array is larger, it is average/worst case O(n) in space complexity. \n",
    "\n",
    "The msBWT is a more sophisticated algorithm with compression features. \n",
    "1328 VS 10000\n",
    "\n",
    "Lecture 13\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 15,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Problem #15.** (5 points) Which in-class exercise was a trap?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 15,
    "section": "answer"
   },
   "source": [
    "Exercise #8"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "number": 16,
    "section": "problem"
   },
   "source": [
    "---\n",
    "**Last Call:** If you have any remaining issues with problem set grades describe them in the following cell. Indicate clearly the *problem set* and the *problem number* for each issue."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "number": 16,
    "section": "answer"
   },
   "source": [
    "No issues\n",
    "\n",
    "Thanks for the semester :) "
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
